{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "whatshap phase --indels --max-coverage 15 --ignore-read-groups -o HuRef.chr22.phased.vcf HuRef_SNPs.22.final.vcf HuRef.chr22.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "HapCUT2/build/extractHAIRS --bam whatshap/HuRef.chr22.bam --VCF whatshap/HuRef_SNPs.22.final.vcf --out whatshap/HuRef_SNPs.22.fragment_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "HapCUT2/build/HAPCUT2 --fragments whatshap/HuRef_SNPs.22.fragment_file --vcf whatshap/HuRef_SNPs.22.final.vcf --out whatshap/HuRef.chr22.phased.HAPCUT2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "whatshap hapcut2vcf -o whatshap/HuRef.chr22.phased.HAPCUT2.vcf whatshap/HuRef_SNPs.22.final.vcf whatshap/HuRef.chr22.phased.HAPCUT2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i in range(1,23):\n",
    "    \n",
    "    print (i)\n",
    "    \n",
    "    os.system('HapCUT2/build/extractHAIRS --bam whatshap/HuRef.chr'+str(i)+'.bam '+ \\\n",
    "        '--VCF whatshap/HuRef_SNPs.'+str(i)+'.final.vcf '+ \\\n",
    "        '--out whatshap/HuRef_SNPs.'+str(i)+'.fragment_file')\n",
    "    \n",
    "    os.system('HapCUT2/build/HAPCUT2 --fragments whatshap/HuRef_SNPs.'+str(i)+'.fragment_file '+ \\\n",
    "        '--vcf whatshap/HuRef_SNPs.'+str(i)+'.final.vcf '+ \\\n",
    "        '--out whatshap/HuRef.chr'+str(i)+'.phased.HAPCUT2.txt')\n",
    "    \n",
    "    os.system('whatshap hapcut2vcf -o whatshap/HuRef.chr'+str(i)+'.phased.HAPCUT2.vcf '+ \n",
    "        'whatshap/HuRef_SNPs.'+str(i)+'.final.vcf ' + \\\n",
    "        'whatshap/HuRef.chr'+str(i)+'.phased.HAPCUT2.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "fls = glob.glob('whatshap/*.phased.vcf')\n",
    "\n",
    "#from fnmatch import fnmatch\n",
    "#[fl for fl in os.listdir('whatshap/.') if fnmatch(fl, '*.phased.vcf')]\n",
    "\n",
    "for fl in fls:\n",
    "    prefix = '.'.join(fl.split('.')[:-1])\n",
    "    os.rename(fl, prefix+'.whatshap.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import OrderedDict\n",
    "hap = OrderedDict()\n",
    "\n",
    "with open('whatshap/Phased_haplotypes.final.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        \n",
    "        hap[lst[0]+':'+lst[1]] = lst[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "metadata": {},
   "outputs": [],
   "source": [
    "### recode\n",
    "\n",
    "fls = glob.glob('whatshap/*.phased.whatshap.vcf')\n",
    "\n",
    "for fl in fls:\n",
    "    prefix = '.'.join(fl.split('.')[:-1])\n",
    "    \n",
    "    fh = open(prefix+'.recode.vcf', 'wt')\n",
    "    fh.write('\\t'.join(['chr','pos','hap1','hap2','block'])+'\\n')\n",
    "\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            pos = lst[0]+':'+lst[1]\n",
    "        \n",
    "            if lst[-1] == '0/1':\n",
    "                fh.write(lst[0]+'\\t'+lst[1]+'\\tNA\\tNA\\tNA\\n')\n",
    "            \n",
    "            else:\n",
    "                block = lst[-1].split(':')[-1]\n",
    "                hap1 = lst[-1].split(':')[0].split('|')[0]\n",
    "            \n",
    "                if hap1 == hap[pos]:\n",
    "                    fh.write('\\t'.join(lst[:2]+['1','0',block])+'\\n')\n",
    "                else:\n",
    "                    fh.write('\\t'.join(lst[:2]+['0','1',block])+'\\n')\n",
    "                \n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "hapBlock = OrderedDict()\n",
    "hapLen = OrderedDict()\n",
    "tot = -1\n",
    "\n",
    "with open('whatshap/HuRef.chr4.phased.whatshap.recode.vcf') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        \n",
    "        tot += 1\n",
    "        \n",
    "        if lst[-1] == 'NA' or lst[-1] == 'block':\n",
    "            continue\n",
    "        \n",
    "        if lst[-1] not in hapBlock.keys():\n",
    "            hapBlock[lst[-1]] = []\n",
    "            hapLen[lst[-1]] = []\n",
    "        hapBlock[lst[-1]].append(lst[2])\n",
    "        hapLen[lst[-1]].append(int(lst[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "130152"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'3793818'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-115-bad51eabce1a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mhapBlock\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'3793818'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m: '3793818'"
     ]
    }
   ],
   "source": [
    "hapBlock['3793818']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(hapBlock['3793818']).most_common(1)[0][1]/len(hapBlock['3793818'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(['0', '0', '0', '0', '0', '1', '1']).most_common(2)[1][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "17\n",
      "18\n",
      "19\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "35\n",
      "36\n",
      "37\n",
      "39\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "76\n",
      "77\n",
      "82\n",
      "83\n",
      "84\n",
      "86\n",
      "88\n",
      "89\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "98\n",
      "99\n",
      "100\n",
      "102\n",
      "103\n",
      "105\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "143\n",
      "144\n",
      "146\n",
      "147\n",
      "150\n",
      "151\n",
      "153\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "170\n",
      "172\n",
      "173\n",
      "174\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "182\n",
      "183\n",
      "188\n",
      "189\n",
      "190\n",
      "195\n",
      "196\n",
      "1049\n",
      "1052\n",
      "1053\n",
      "1054\n",
      "1056\n",
      "1057\n",
      "1058\n",
      "1060\n",
      "1062\n",
      "1063\n",
      "1064\n",
      "1066\n",
      "1068\n",
      "1075\n",
      "1076\n",
      "1079\n",
      "1081\n",
      "1082\n",
      "1084\n",
      "1086\n",
      "1087\n",
      "1088\n",
      "1089\n",
      "1090\n",
      "1092\n",
      "1093\n",
      "1094\n",
      "1095\n",
      "1096\n",
      "1097\n",
      "1104\n",
      "1105\n",
      "1115\n",
      "1354\n",
      "1418\n",
      "1419\n",
      "1421\n",
      "1423\n",
      "1425\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "cmplt = 0\n",
    "i = 0\n",
    "for ke, val in hapBlock.items():\n",
    "    cmplt += len(val)\n",
    "    ratio = Counter(val).most_common(1)[0][1]/len(val)\n",
    "    if (ratio < 1):\n",
    "        i += Counter(val).most_common(2)[1][1]\n",
    "        print (i)\n",
    "        \n",
    "        #i += 1\n",
    "        #print (ke+'\\t'+str(ratio))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24248"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(hapBlock)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "331"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "85015"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmplt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.005789392537773667"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmplt/tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxBlock = 0\n",
    "\n",
    "for ke, val in hapLen.items():\n",
    "    l = val[-1] - val[0]\n",
    "    \n",
    "    if l > maxBlock:\n",
    "        maxBlock = l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "98.861243"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "maxBlock/1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6553731113166821"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "cmplt/tot\n",
    "\n",
    "\n",
    "\n",
    "### Completeness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Switch error rate (SER)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "with open('Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr4.haplotype.all.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split()\n",
    "        \n",
    "        if lst[1] == '1':\n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "513"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['whatshap/HuRef.chr2.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr16.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr1.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr19.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr17.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr10.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr11.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr3.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr12.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr6.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr22.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr7.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr8.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr18.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr21.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr9.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr13.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr4.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr14.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr5.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr20.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr15.phased.whatshap.recode.vcf']"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fls = glob.glob('whatshap/*.phased.whatshap.recode.vcf')\n",
    "fls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['whatshap/HuRef.chr1.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr2.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr3.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr4.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr5.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr6.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr7.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr8.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr9.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr10.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr11.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr12.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr13.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr14.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr15.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr16.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr17.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr18.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr19.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr20.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr21.phased.whatshap.recode.vcf',\n",
       " 'whatshap/HuRef.chr22.phased.whatshap.recode.vcf']"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(fls, key = lambda x:int(x.split('.')[1].replace('chr', '')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "####################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "102324460\n",
      "158489704\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 37215\n",
      "Percentage of phased SNVs: 28.689\n",
      "Number of phased blocks: 13510\n",
      "Completeness (NC): 18.274\n",
      "Number of switch errors: 384\n",
      "Switch error rate: 1.620000\n",
      "Blockwise Hamming distance: 369\n",
      "Blockwise Hamming rate: 0.992000\n",
      "Largest haplotype segment: 143.915M\n",
      "Most-variants-phased (MVP) segment: 103\n",
      "\n",
      "=======================chr2=======================\n",
      "63154816\n",
      "117204657\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 40655\n",
      "Percentage of phased SNVs: 28.697\n",
      "Number of phased blocks: 14867\n",
      "Completeness (NC): 18.203\n",
      "Number of switch errors: 325\n",
      "Switch error rate: 1.260000\n",
      "Blockwise Hamming distance: 326\n",
      "Blockwise Hamming rate: 0.802000\n",
      "Largest haplotype segment: 118.066M\n",
      "Most-variants-phased (MVP) segment: 77\n",
      "\n",
      "=======================chr3=======================\n",
      "53471338\n",
      "103929183\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 38953\n",
      "Percentage of phased SNVs: 30.612\n",
      "Number of phased blocks: 13968\n",
      "Completeness (NC): 19.635\n",
      "Number of switch errors: 187\n",
      "Switch error rate: 0.748000\n",
      "Blockwise Hamming distance: 192\n",
      "Blockwise Hamming rate: 0.493000\n",
      "Largest haplotype segment: 115.9M\n",
      "Most-variants-phased (MVP) segment: 121\n",
      "\n",
      "=======================chr4=======================\n",
      "21465757\n",
      "162007987\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 39952\n",
      "Percentage of phased SNVs: 30.696\n",
      "Number of phased blocks: 14482\n",
      "Completeness (NC): 19.569\n",
      "Number of switch errors: 276\n",
      "Switch error rate: 1.084000\n",
      "Blockwise Hamming distance: 620\n",
      "Blockwise Hamming rate: 1.552000\n",
      "Largest haplotype segment: 59.305M\n",
      "Most-variants-phased (MVP) segment: 950\n",
      "\n",
      "=======================chr5=======================\n",
      "5876623\n",
      "21523395\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 35374\n",
      "Percentage of phased SNVs: 30.763\n",
      "Number of phased blocks: 12716\n",
      "Completeness (NC): 19.705\n",
      "Number of switch errors: 237\n",
      "Switch error rate: 1.046000\n",
      "Blockwise Hamming distance: 315\n",
      "Blockwise Hamming rate: 0.890000\n",
      "Largest haplotype segment: 92.028M\n",
      "Most-variants-phased (MVP) segment: 321\n",
      "\n",
      "=======================chr6=======================\n",
      "32527025\n",
      "57302641\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 38563\n",
      "Percentage of phased SNVs: 33.874\n",
      "Number of phased blocks: 12108\n",
      "Completeness (NC): 23.238\n",
      "Number of switch errors: 837\n",
      "Switch error rate: 3.164000\n",
      "Blockwise Hamming distance: 1292\n",
      "Blockwise Hamming rate: 3.350000\n",
      "Largest haplotype segment: 64.495M\n",
      "Most-variants-phased (MVP) segment: 1366\n",
      "\n",
      "=======================chr7=======================\n",
      "40352072\n",
      "12012587\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 29102\n",
      "Percentage of phased SNVs: 29.938\n",
      "Number of phased blocks: 10395\n",
      "Completeness (NC): 19.244\n",
      "Number of switch errors: 146\n",
      "Switch error rate: 0.780000\n",
      "Blockwise Hamming distance: 172\n",
      "Blockwise Hamming rate: 0.591000\n",
      "Largest haplotype segment: 88.848M\n",
      "Most-variants-phased (MVP) segment: 66\n",
      "\n",
      "=======================chr8=======================\n",
      "55692112\n",
      "3998147\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 30804\n",
      "Percentage of phased SNVs: 32.218\n",
      "Number of phased blocks: 10801\n",
      "Completeness (NC): 20.921\n",
      "Number of switch errors: 109\n",
      "Switch error rate: 0.545000\n",
      "Blockwise Hamming distance: 118\n",
      "Blockwise Hamming rate: 0.383000\n",
      "Largest haplotype segment: 82.309M\n",
      "Most-variants-phased (MVP) segment: 71\n",
      "\n",
      "=======================chr9=======================\n",
      "4423007\n",
      "7522561\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 21049\n",
      "Percentage of phased SNVs: 30.341\n",
      "Number of phased blocks: 7663\n",
      "Completeness (NC): 19.295\n",
      "Number of switch errors: 115\n",
      "Switch error rate: 0.859000\n",
      "Blockwise Hamming distance: 109\n",
      "Blockwise Hamming rate: 0.518000\n",
      "Largest haplotype segment: 103.014M\n",
      "Most-variants-phased (MVP) segment: 57\n",
      "\n",
      "======================chr10=======================\n",
      "9094997\n",
      "117337166\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 26687\n",
      "Percentage of phased SNVs: 30.035\n",
      "Number of phased blocks: 9738\n",
      "Completeness (NC): 19.075\n",
      "Number of switch errors: 155\n",
      "Switch error rate: 0.915000\n",
      "Blockwise Hamming distance: 143\n",
      "Blockwise Hamming rate: 0.536000\n",
      "Largest haplotype segment: 99.92M\n",
      "Most-variants-phased (MVP) segment: 56\n",
      "\n",
      "======================chr11=======================\n",
      "23123811\n",
      "26227266\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 26419\n",
      "Percentage of phased SNVs: 30.522\n",
      "Number of phased blocks: 9336\n",
      "Completeness (NC): 19.736\n",
      "Number of switch errors: 124\n",
      "Switch error rate: 0.726000\n",
      "Blockwise Hamming distance: 124\n",
      "Blockwise Hamming rate: 0.469000\n",
      "Largest haplotype segment: 70.495M\n",
      "Most-variants-phased (MVP) segment: 76\n",
      "\n",
      "======================chr12=======================\n",
      "59839867\n",
      "46474068\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 21078\n",
      "Percentage of phased SNVs: 27.073\n",
      "Number of phased blocks: 8056\n",
      "Completeness (NC): 16.726\n",
      "Number of switch errors: 62\n",
      "Switch error rate: 0.476000\n",
      "Blockwise Hamming distance: 64\n",
      "Blockwise Hamming rate: 0.304000\n",
      "Largest haplotype segment: 55.809M\n",
      "Most-variants-phased (MVP) segment: 43\n",
      "\n",
      "======================chr13=======================\n",
      "22896139\n",
      "63628684\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 20505\n",
      "Percentage of phased SNVs: 30.852\n",
      "Number of phased blocks: 7440\n",
      "Completeness (NC): 19.658\n",
      "Number of switch errors: 94\n",
      "Switch error rate: 0.719000\n",
      "Blockwise Hamming distance: 84\n",
      "Blockwise Hamming rate: 0.410000\n",
      "Largest haplotype segment: 80.755M\n",
      "Most-variants-phased (MVP) segment: 195\n",
      "\n",
      "======================chr14=======================\n",
      "26778722\n",
      "39140187\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 14967\n",
      "Percentage of phased SNVs: 28.825\n",
      "Number of phased blocks: 5452\n",
      "Completeness (NC): 18.325\n",
      "Number of switch errors: 72\n",
      "Switch error rate: 0.757000\n",
      "Blockwise Hamming distance: 76\n",
      "Blockwise Hamming rate: 0.508000\n",
      "Largest haplotype segment: 69.324M\n",
      "Most-variants-phased (MVP) segment: 45\n",
      "\n",
      "======================chr15=======================\n",
      "50326234\n",
      "22410004\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 13255\n",
      "Percentage of phased SNVs: 27.634\n",
      "Number of phased blocks: 4985\n",
      "Completeness (NC): 17.241\n",
      "Number of switch errors: 156\n",
      "Switch error rate: 1.886000\n",
      "Blockwise Hamming distance: 144\n",
      "Blockwise Hamming rate: 1.086000\n",
      "Largest haplotype segment: 52.047M\n",
      "Most-variants-phased (MVP) segment: 35\n",
      "\n",
      "======================chr16=======================\n",
      "13726154\n",
      "5731033\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 13427\n",
      "Percentage of phased SNVs: 29.322\n",
      "Number of phased blocks: 4670\n",
      "Completeness (NC): 19.123\n",
      "Number of switch errors: 194\n",
      "Switch error rate: 2.215000\n",
      "Blockwise Hamming distance: 191\n",
      "Blockwise Hamming rate: 1.423000\n",
      "Largest haplotype segment: 66.405M\n",
      "Most-variants-phased (MVP) segment: 181\n",
      "\n",
      "======================chr17=======================\n",
      "32403029\n",
      "3111798\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 8691\n",
      "Percentage of phased SNVs: 23.592\n",
      "Number of phased blocks: 3278\n",
      "Completeness (NC): 14.694\n",
      "Number of switch errors: 137\n",
      "Switch error rate: 2.531000\n",
      "Blockwise Hamming distance: 127\n",
      "Blockwise Hamming rate: 1.461000\n",
      "Largest haplotype segment: 22.688M\n",
      "Most-variants-phased (MVP) segment: 24\n",
      "\n",
      "======================chr18=======================\n",
      "1386741\n",
      "69993799\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 12973\n",
      "Percentage of phased SNVs: 28.362\n",
      "Number of phased blocks: 4850\n",
      "Completeness (NC): 17.759\n",
      "Number of switch errors: 43\n",
      "Switch error rate: 0.529000\n",
      "Blockwise Hamming distance: 46\n",
      "Blockwise Hamming rate: 0.355000\n",
      "Largest haplotype segment: 67.326M\n",
      "Most-variants-phased (MVP) segment: 78\n",
      "\n",
      "======================chr19=======================\n",
      "53186948\n",
      "20751225\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 5556\n",
      "Percentage of phased SNVs: 22.819\n",
      "Number of phased blocks: 2066\n",
      "Completeness (NC): 14.334\n",
      "Number of switch errors: 14\n",
      "Switch error rate: 0.401000\n",
      "Blockwise Hamming distance: 14\n",
      "Blockwise Hamming rate: 0.252000\n",
      "Largest haplotype segment: 0.295M\n",
      "Most-variants-phased (MVP) segment: 58\n",
      "\n",
      "======================chr20=======================\n",
      "9926657\n",
      "26205889\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 7861\n",
      "Percentage of phased SNVs: 23.949\n",
      "Number of phased blocks: 3041\n",
      "Completeness (NC): 14.684\n",
      "Number of switch errors: 81\n",
      "Switch error rate: 1.680000\n",
      "Blockwise Hamming distance: 88\n",
      "Blockwise Hamming rate: 1.119000\n",
      "Largest haplotype segment: 9.493M\n",
      "Most-variants-phased (MVP) segment: 39\n",
      "\n",
      "======================chr21=======================\n",
      "23750408\n",
      "11042288\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 7325\n",
      "Percentage of phased SNVs: 32.889\n",
      "Number of phased blocks: 2473\n",
      "Completeness (NC): 21.785\n",
      "Number of switch errors: 221\n",
      "Switch error rate: 4.555000\n",
      "Blockwise Hamming distance: 197\n",
      "Blockwise Hamming rate: 2.689000\n",
      "Largest haplotype segment: 0.034M\n",
      "Most-variants-phased (MVP) segment: 141\n",
      "\n",
      "======================chr22=======================\n",
      "22697358\n",
      "35330065\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 3251\n",
      "Percentage of phased SNVs: 21.193\n",
      "Number of phased blocks: 1239\n",
      "Completeness (NC): 13.116\n",
      "Number of switch errors: 9\n",
      "Switch error rate: 0.447000\n",
      "Blockwise Hamming distance: 9\n",
      "Blockwise Hamming rate: 0.277000\n",
      "Largest haplotype segment: 0.016M\n",
      "Most-variants-phased (MVP) segment: 17\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('whatshap/*.3.phased.whatshap.recode.vcf')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[1].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[1]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = OrderedDict()\n",
    "    hapPos = OrderedDict()\n",
    "    numberOfSNPs = -1\n",
    "\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            numberOfSNPs += 1\n",
    "        \n",
    "            if lst[-1] == 'NA' or lst[-1] == 'block':\n",
    "                continue\n",
    "        \n",
    "            if lst[-1] not in hapBlock.keys():\n",
    "                hapBlock[lst[-1]] = []\n",
    "                hapPos[lst[-1]] = []\n",
    "            hapBlock[lst[-1]].append(lst[2])\n",
    "            hapPos[lst[-1]].append(int(lst[1]))\n",
    "    \n",
    "    \n",
    "    numberOfSegments = len(hapBlock)\n",
    "    \n",
    "    phasedSNPs = 0\n",
    "    \n",
    "    mismatch = switchError = hammingDistance = 0\n",
    "    \n",
    "    #hapChr = []\n",
    "    for ke, val in hapBlock.items():\n",
    "        phasedSNPs += len(val)\n",
    "        \n",
    "        if len(Counter(val))==2:\n",
    "            hammingDistance += Counter(val).most_common(2)[1][1]\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        #hapChr.append(len(val))\n",
    "        \n",
    "        v1 = val[1:]\n",
    "        v2 = val[:-1]\n",
    "        \n",
    "        vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "        switchError += sum(vd)\n",
    "\n",
    "        if len(vd)==1:\n",
    "            continue\n",
    "                \n",
    "        indexV = [i+1 for i in range(len(vd)) if vd[i]==True]\n",
    "        indexV1 = indexV[1:]\n",
    "        indexV2 = indexV[:-1]\n",
    "        \n",
    "        indexVd = [a - b for a, b in zip(indexV1,indexV2)]\n",
    "        \n",
    "        mismatch += indexVd.count(1)\n",
    "        \n",
    "        #indexV1.extend([len(val)])\n",
    "        #indexV2.insert(0,0)\n",
    "        #hapChr.extend([a - b for a, b in zip(indexV1, indexV2)])\n",
    "        \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs*100,3)\n",
    "        \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "    completeness = round((phasedSNPs-numberOfSegments)/numberOfSNPs*100,3)\n",
    "    \n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    #v1 = hapChr[1:]\n",
    "    #v2 = hapChr[:-1]\n",
    "    #error = sum([a != b for a, b in zip(v1,v2)])\n",
    "    #ser = error/(cmplt)*100\n",
    "    \n",
    "    \n",
    "    LHSLength = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = val[-1] - val[0] + 1\n",
    "            \n",
    "        if l > LHSLength:\n",
    "            LHSLength = l\n",
    "            LHSVariant = len(val)\n",
    "            \n",
    "    LHSLength = round(LHSLength/1000000,3)\n",
    "        \n",
    "    for ke, val in hapPos.items():\n",
    "        l = round((val[-1] - val[0] + 1)/1000000,3)\n",
    "        \n",
    "        if l == LHSLength:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    MVP = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = len(val)\n",
    "            \n",
    "        if l > MVP:\n",
    "            MVP = l\n",
    "    \n",
    "    for ke, val in hapPos.items():\n",
    "        if len(val) == MVP:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    #QAN50 = 0\n",
    "    #for size in sorted(hapChr, reverse=True):\n",
    "    #    QAN50 += size\n",
    "        #print ('largest %s, %s' %(size, QAN50))\n",
    "    \n",
    "    \n",
    "        #if QAN50 > phased/2:\n",
    "            #QAN50 = size\n",
    "            #print ('QAN %s' %size)\n",
    "            #break\n",
    "    \n",
    "        #ratio = Counter(val).most_common(1)[0][1]/len(val)\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        \n",
    "    metrics[chrom] = [numberOfSNPs, phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs)\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = open('whatshap/metrics.whatshap.3.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "14320465\n",
      "158497079\n",
      "158503221\n",
      "14320465\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 35224\n",
      "Percentage of phased SNVs: 27.154\n",
      "Number of phased blocks: 13049\n",
      "Completeness (NC): 17.095\n",
      "Number of switch errors: 272\n",
      "Switch error rate: 1.227000\n",
      "Blockwise Hamming distance: 238\n",
      "Blockwise Hamming rate: 0.676000\n",
      "Largest haplotype segment: 0.005M\n",
      "Most-variants-phased (MVP) segment: 44\n",
      "\n",
      "=======================chr2=======================\n",
      "133047563\n",
      "133047563\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 38598\n",
      "Percentage of phased SNVs: 27.245\n",
      "Number of phased blocks: 14427\n",
      "Completeness (NC): 17.062\n",
      "Number of switch errors: 246\n",
      "Switch error rate: 1.018000\n",
      "Blockwise Hamming distance: 246\n",
      "Blockwise Hamming rate: 0.637000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 63\n",
      "\n",
      "=======================chr3=======================\n",
      "103937539\n",
      "103937539\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 36968\n",
      "Percentage of phased SNVs: 29.052\n",
      "Number of phased blocks: 13578\n",
      "Completeness (NC): 18.382\n",
      "Number of switch errors: 113\n",
      "Switch error rate: 0.483000\n",
      "Blockwise Hamming distance: 111\n",
      "Blockwise Hamming rate: 0.300000\n",
      "Largest haplotype segment: 0.01M\n",
      "Most-variants-phased (MVP) segment: 76\n",
      "\n",
      "=======================chr4=======================\n",
      "190643784\n",
      "190667406\n",
      "190667406\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 38573\n",
      "Percentage of phased SNVs: 29.637\n",
      "Number of phased blocks: 14109\n",
      "Completeness (NC): 18.796\n",
      "Number of switch errors: 221\n",
      "Switch error rate: 0.903000\n",
      "Blockwise Hamming distance: 203\n",
      "Blockwise Hamming rate: 0.526000\n",
      "Largest haplotype segment: 0.013M\n",
      "Most-variants-phased (MVP) segment: 209\n",
      "\n",
      "=======================chr5=======================\n",
      "8015575\n",
      "21547447\n",
      "21547447\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 33615\n",
      "Percentage of phased SNVs: 29.233\n",
      "Number of phased blocks: 12293\n",
      "Completeness (NC): 18.543\n",
      "Number of switch errors: 176\n",
      "Switch error rate: 0.825000\n",
      "Blockwise Hamming distance: 209\n",
      "Blockwise Hamming rate: 0.622000\n",
      "Largest haplotype segment: 0.007M\n",
      "Most-variants-phased (MVP) segment: 87\n",
      "\n",
      "=======================chr6=======================\n",
      "57472841\n",
      "57472841\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 36851\n",
      "Percentage of phased SNVs: 32.37\n",
      "Number of phased blocks: 11879\n",
      "Completeness (NC): 21.935\n",
      "Number of switch errors: 741\n",
      "Switch error rate: 2.967000\n",
      "Blockwise Hamming distance: 750\n",
      "Blockwise Hamming rate: 2.035000\n",
      "Largest haplotype segment: 0.024M\n",
      "Most-variants-phased (MVP) segment: 254\n",
      "\n",
      "=======================chr7=======================\n",
      "66665292\n",
      "4071753\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 27662\n",
      "Percentage of phased SNVs: 28.457\n",
      "Number of phased blocks: 10068\n",
      "Completeness (NC): 18.099\n",
      "Number of switch errors: 79\n",
      "Switch error rate: 0.449000\n",
      "Blockwise Hamming distance: 74\n",
      "Blockwise Hamming rate: 0.268000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 51\n",
      "\n",
      "=======================chr8=======================\n",
      "3222789\n",
      "5001614\n",
      "3998147\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 29311\n",
      "Percentage of phased SNVs: 30.656\n",
      "Number of phased blocks: 10518\n",
      "Completeness (NC): 19.655\n",
      "Number of switch errors: 53\n",
      "Switch error rate: 0.282000\n",
      "Blockwise Hamming distance: 43\n",
      "Blockwise Hamming rate: 0.147000\n",
      "Largest haplotype segment: 0.005M\n",
      "Most-variants-phased (MVP) segment: 71\n",
      "\n",
      "=======================chr9=======================\n",
      "7524011\n",
      "7524011\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 20013\n",
      "Percentage of phased SNVs: 28.848\n",
      "Number of phased blocks: 7442\n",
      "Completeness (NC): 18.121\n",
      "Number of switch errors: 75\n",
      "Switch error rate: 0.597000\n",
      "Blockwise Hamming distance: 67\n",
      "Blockwise Hamming rate: 0.335000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 41\n",
      "\n",
      "======================chr10=======================\n",
      "95592251\n",
      "26355227\n",
      "95592251\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 25654\n",
      "Percentage of phased SNVs: 28.872\n",
      "Number of phased blocks: 9546\n",
      "Completeness (NC): 18.129\n",
      "Number of switch errors: 104\n",
      "Switch error rate: 0.646000\n",
      "Blockwise Hamming distance: 94\n",
      "Blockwise Hamming rate: 0.366000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 38\n",
      "\n",
      "======================chr11=======================\n",
      "24564217\n",
      "92289781\n",
      "99272344\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 25138\n",
      "Percentage of phased SNVs: 29.042\n",
      "Number of phased blocks: 9072\n",
      "Completeness (NC): 18.561\n",
      "Number of switch errors: 59\n",
      "Switch error rate: 0.367000\n",
      "Blockwise Hamming distance: 56\n",
      "Blockwise Hamming rate: 0.223000\n",
      "Largest haplotype segment: 0.007M\n",
      "Most-variants-phased (MVP) segment: 43\n",
      "\n",
      "======================chr12=======================\n",
      "61558921\n",
      "61558921\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 20211\n",
      "Percentage of phased SNVs: 25.959\n",
      "Number of phased blocks: 7827\n",
      "Completeness (NC): 15.906\n",
      "Number of switch errors: 29\n",
      "Switch error rate: 0.234000\n",
      "Blockwise Hamming distance: 32\n",
      "Blockwise Hamming rate: 0.158000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 26\n",
      "\n",
      "======================chr13=======================\n",
      "63628684\n",
      "63628684\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 19502\n",
      "Percentage of phased SNVs: 29.343\n",
      "Number of phased blocks: 7215\n",
      "Completeness (NC): 18.487\n",
      "Number of switch errors: 55\n",
      "Switch error rate: 0.448000\n",
      "Blockwise Hamming distance: 45\n",
      "Blockwise Hamming rate: 0.231000\n",
      "Largest haplotype segment: 0.007M\n",
      "Most-variants-phased (MVP) segment: 97\n",
      "\n",
      "======================chr14=======================\n",
      "39140187\n",
      "39146985\n",
      "55910679\n",
      "63427586\n",
      "87736137\n",
      "87736137\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 14248\n",
      "Percentage of phased SNVs: 27.44\n",
      "Number of phased blocks: 5327\n",
      "Completeness (NC): 17.181\n",
      "Number of switch errors: 34\n",
      "Switch error rate: 0.381000\n",
      "Blockwise Hamming distance: 35\n",
      "Blockwise Hamming rate: 0.246000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 29\n",
      "\n",
      "======================chr15=======================\n",
      "22414580\n",
      "54773691\n",
      "55142630\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 12719\n",
      "Percentage of phased SNVs: 26.517\n",
      "Number of phased blocks: 4834\n",
      "Completeness (NC): 16.439\n",
      "Number of switch errors: 127\n",
      "Switch error rate: 1.611000\n",
      "Blockwise Hamming distance: 115\n",
      "Blockwise Hamming rate: 0.904000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 23\n",
      "\n",
      "======================chr16=======================\n",
      "76444155\n",
      "76444155\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 12737\n",
      "Percentage of phased SNVs: 27.815\n",
      "Number of phased blocks: 4519\n",
      "Completeness (NC): 17.946\n",
      "Number of switch errors: 159\n",
      "Switch error rate: 1.935000\n",
      "Blockwise Hamming distance: 157\n",
      "Blockwise Hamming rate: 1.233000\n",
      "Largest haplotype segment: 0.008M\n",
      "Most-variants-phased (MVP) segment: 65\n",
      "\n",
      "======================chr17=======================\n",
      "14067760\n",
      "39309694\n",
      "39440801\n",
      "50008688\n",
      "50339911\n",
      "50808933\n",
      "53734571\n",
      "63903880\n",
      "3111798\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 8125\n",
      "Percentage of phased SNVs: 22.056\n",
      "Number of phased blocks: 3114\n",
      "Completeness (NC): 13.603\n",
      "Number of switch errors: 111\n",
      "Switch error rate: 2.215000\n",
      "Blockwise Hamming distance: 101\n",
      "Blockwise Hamming rate: 1.243000\n",
      "Largest haplotype segment: 0.002M\n",
      "Most-variants-phased (MVP) segment: 24\n",
      "\n",
      "======================chr18=======================\n",
      "70000592\n",
      "70000592\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 12298\n",
      "Percentage of phased SNVs: 26.886\n",
      "Number of phased blocks: 4680\n",
      "Completeness (NC): 16.655\n",
      "Number of switch errors: 16\n",
      "Switch error rate: 0.210000\n",
      "Blockwise Hamming distance: 16\n",
      "Blockwise Hamming rate: 0.130000\n",
      "Largest haplotype segment: 0.005M\n",
      "Most-variants-phased (MVP) segment: 49\n",
      "\n",
      "======================chr19=======================\n",
      "51774129\n",
      "20752051\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 5309\n",
      "Percentage of phased SNVs: 21.805\n",
      "Number of phased blocks: 1993\n",
      "Completeness (NC): 13.619\n",
      "Number of switch errors: 11\n",
      "Switch error rate: 0.332000\n",
      "Blockwise Hamming distance: 11\n",
      "Blockwise Hamming rate: 0.207000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 41\n",
      "\n",
      "======================chr20=======================\n",
      "26206229\n",
      "26206229\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 7565\n",
      "Percentage of phased SNVs: 23.047\n",
      "Number of phased blocks: 2935\n",
      "Completeness (NC): 14.106\n",
      "Number of switch errors: 64\n",
      "Switch error rate: 1.382000\n",
      "Blockwise Hamming distance: 73\n",
      "Blockwise Hamming rate: 0.965000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 38\n",
      "\n",
      "======================chr21=======================\n",
      "29564744\n",
      "11052183\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 6934\n",
      "Percentage of phased SNVs: 31.133\n",
      "Number of phased blocks: 2383\n",
      "Completeness (NC): 20.434\n",
      "Number of switch errors: 191\n",
      "Switch error rate: 4.197000\n",
      "Blockwise Hamming distance: 173\n",
      "Blockwise Hamming rate: 2.495000\n",
      "Largest haplotype segment: 0.008M\n",
      "Most-variants-phased (MVP) segment: 65\n",
      "\n",
      "======================chr22=======================\n",
      "35331402\n",
      "35469757\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 3014\n",
      "Percentage of phased SNVs: 19.648\n",
      "Number of phased blocks: 1157\n",
      "Completeness (NC): 12.106\n",
      "Number of switch errors: 3\n",
      "Switch error rate: 0.162000\n",
      "Blockwise Hamming distance: 3\n",
      "Blockwise Hamming rate: 0.100000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 16\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('whatshap/*.3.phased.HAPCUT2.recode.vcf')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[1].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[1]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = OrderedDict()\n",
    "    hapPos = OrderedDict()\n",
    "    numberOfSNPs = -1\n",
    "\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            numberOfSNPs += 1\n",
    "        \n",
    "            if lst[-1] == 'NA' or lst[-1] == 'block':\n",
    "                continue\n",
    "        \n",
    "            if lst[-1] not in hapBlock.keys():\n",
    "                hapBlock[lst[-1]] = []\n",
    "                hapPos[lst[-1]] = []\n",
    "            hapBlock[lst[-1]].append(lst[2])\n",
    "            hapPos[lst[-1]].append(int(lst[1]))\n",
    "    \n",
    "    \n",
    "    numberOfSegments = len(hapBlock)\n",
    "    \n",
    "    phasedSNPs = 0\n",
    "    \n",
    "    mismatch = switchError = hammingDistance = 0\n",
    "    \n",
    "    #hapChr = []\n",
    "    for ke, val in hapBlock.items():\n",
    "        phasedSNPs += len(val)\n",
    "        \n",
    "        if len(Counter(val))==2:\n",
    "            hammingDistance += Counter(val).most_common(2)[1][1]\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        #hapChr.append(len(val))\n",
    "        \n",
    "        v1 = val[1:]\n",
    "        v2 = val[:-1]\n",
    "        \n",
    "        vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "        switchError += sum(vd)\n",
    "\n",
    "        if len(vd)==1:\n",
    "            continue\n",
    "                \n",
    "        indexV = [i+1 for i in range(len(vd)) if vd[i]==True]\n",
    "        indexV1 = indexV[1:]\n",
    "        indexV2 = indexV[:-1]\n",
    "        \n",
    "        indexVd = [a - b for a, b in zip(indexV1,indexV2)]\n",
    "        \n",
    "        mismatch += indexVd.count(1)\n",
    "        \n",
    "        #indexV1.extend([len(val)])\n",
    "        #indexV2.insert(0,0)\n",
    "        #hapChr.extend([a - b for a, b in zip(indexV1, indexV2)])\n",
    "        \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs*100,3)\n",
    "        \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "    completeness = round((phasedSNPs-numberOfSegments)/numberOfSNPs*100,3)\n",
    "    \n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    #v1 = hapChr[1:]\n",
    "    #v2 = hapChr[:-1]\n",
    "    #error = sum([a != b for a, b in zip(v1,v2)])\n",
    "    #ser = error/(cmplt)*100\n",
    "    \n",
    "    \n",
    "    LHSLength = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = val[-1] - val[0] + 1\n",
    "            \n",
    "        if l > LHSLength:\n",
    "            LHSLength = l\n",
    "            LHSVariant = len(val)\n",
    "            \n",
    "    LHSLength = round(LHSLength/1000000,3)\n",
    "        \n",
    "    for ke, val in hapPos.items():\n",
    "        l = round((val[-1] - val[0] + 1)/1000000,3)\n",
    "        \n",
    "        if l == LHSLength:\n",
    "            print (ke)\n",
    "        \n",
    "        \n",
    "    MVP = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = len(val)\n",
    "            \n",
    "        if l > MVP:\n",
    "            MVP = l\n",
    "            \n",
    "    for ke, val in hapPos.items():\n",
    "        if len(val) == MVP:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    #QAN50 = 0\n",
    "    #for size in sorted(hapChr, reverse=True):\n",
    "    #    QAN50 += size\n",
    "        #print ('largest %s, %s' %(size, QAN50))\n",
    "    \n",
    "    \n",
    "        #if QAN50 > phased/2:\n",
    "            #QAN50 = size\n",
    "            #print ('QAN %s' %size)\n",
    "            #break\n",
    "    \n",
    "        #ratio = Counter(val).most_common(1)[0][1]/len(val)\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        \n",
    "    metrics[chrom] = [numberOfSNPs, phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs)\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fh = open('whatshap/metrics.HAPCUT2.11.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 129241\n",
      "Percentage of phased SNVs: 99.631\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.631\n",
      "Number of switch errors: 511\n",
      "Switch error rate: 0.395000\n",
      "Blockwise Hamming distance: 315\n",
      "Blockwise Hamming rate: 0.244000\n",
      "Largest haplotype segment: 248.393M\n",
      "Most-variants-phased (MVP) segment: 129241\n",
      "\n",
      "=======================chr2=======================\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 141225\n",
      "Percentage of phased SNVs: 99.687\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.687\n",
      "Number of switch errors: 426\n",
      "Switch error rate: 0.302000\n",
      "Blockwise Hamming distance: 307\n",
      "Blockwise Hamming rate: 0.217000\n",
      "Largest haplotype segment: 242.959M\n",
      "Most-variants-phased (MVP) segment: 141225\n",
      "\n",
      "=======================chr3=======================\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 127069\n",
      "Percentage of phased SNVs: 99.86\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.86\n",
      "Number of switch errors: 290\n",
      "Switch error rate: 0.228000\n",
      "Blockwise Hamming distance: 198\n",
      "Blockwise Hamming rate: 0.156000\n",
      "Largest haplotype segment: 197.794M\n",
      "Most-variants-phased (MVP) segment: 127069\n",
      "\n",
      "=======================chr4=======================\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 129995\n",
      "Percentage of phased SNVs: 99.879\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.879\n",
      "Number of switch errors: 420\n",
      "Switch error rate: 0.323000\n",
      "Blockwise Hamming distance: 513\n",
      "Blockwise Hamming rate: 0.395000\n",
      "Largest haplotype segment: 191.009M\n",
      "Most-variants-phased (MVP) segment: 129995\n",
      "\n",
      "=======================chr5=======================\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 114773\n",
      "Percentage of phased SNVs: 99.813\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.813\n",
      "Number of switch errors: 245\n",
      "Switch error rate: 0.213000\n",
      "Blockwise Hamming distance: 165\n",
      "Blockwise Hamming rate: 0.144000\n",
      "Largest haplotype segment: 180.652M\n",
      "Most-variants-phased (MVP) segment: 114773\n",
      "\n",
      "=======================chr6=======================\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 113185\n",
      "Percentage of phased SNVs: 99.421\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.421\n",
      "Number of switch errors: 644\n",
      "Switch error rate: 0.569000\n",
      "Blockwise Hamming distance: 453\n",
      "Blockwise Hamming rate: 0.400000\n",
      "Largest haplotype segment: 170.71M\n",
      "Most-variants-phased (MVP) segment: 113185\n",
      "\n",
      "=======================chr7=======================\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 96828\n",
      "Percentage of phased SNVs: 99.609\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.609\n",
      "Number of switch errors: 228\n",
      "Switch error rate: 0.235000\n",
      "Blockwise Hamming distance: 136\n",
      "Blockwise Hamming rate: 0.140000\n",
      "Largest haplotype segment: 159.066M\n",
      "Most-variants-phased (MVP) segment: 96828\n",
      "\n",
      "=======================chr8=======================\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 95529\n",
      "Percentage of phased SNVs: 99.913\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.913\n",
      "Number of switch errors: 80\n",
      "Switch error rate: 0.084000\n",
      "Blockwise Hamming distance: 42\n",
      "Blockwise Hamming rate: 0.044000\n",
      "Largest haplotype segment: 146.137M\n",
      "Most-variants-phased (MVP) segment: 95529\n",
      "\n",
      "=======================chr9=======================\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 69197\n",
      "Percentage of phased SNVs: 99.745\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.745\n",
      "Number of switch errors: 249\n",
      "Switch error rate: 0.360000\n",
      "Blockwise Hamming distance: 208\n",
      "Blockwise Hamming rate: 0.301000\n",
      "Largest haplotype segment: 140.973M\n",
      "Most-variants-phased (MVP) segment: 69197\n",
      "\n",
      "======================chr10=======================\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 88627\n",
      "Percentage of phased SNVs: 99.745\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.745\n",
      "Number of switch errors: 264\n",
      "Switch error rate: 0.298000\n",
      "Blockwise Hamming distance: 193\n",
      "Blockwise Hamming rate: 0.218000\n",
      "Largest haplotype segment: 135.371M\n",
      "Most-variants-phased (MVP) segment: 88627\n",
      "\n",
      "======================chr11=======================\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 86432\n",
      "Percentage of phased SNVs: 99.856\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.856\n",
      "Number of switch errors: 158\n",
      "Switch error rate: 0.183000\n",
      "Blockwise Hamming distance: 101\n",
      "Blockwise Hamming rate: 0.117000\n",
      "Largest haplotype segment: 134.713M\n",
      "Most-variants-phased (MVP) segment: 86432\n",
      "\n",
      "======================chr12=======================\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 77772\n",
      "Percentage of phased SNVs: 99.892\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.892\n",
      "Number of switch errors: 114\n",
      "Switch error rate: 0.147000\n",
      "Blockwise Hamming distance: 73\n",
      "Blockwise Hamming rate: 0.094000\n",
      "Largest haplotype segment: 133.619M\n",
      "Most-variants-phased (MVP) segment: 77772\n",
      "\n",
      "======================chr13=======================\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 66402\n",
      "Percentage of phased SNVs: 99.908\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.908\n",
      "Number of switch errors: 96\n",
      "Switch error rate: 0.145000\n",
      "Blockwise Hamming distance: 55\n",
      "Blockwise Hamming rate: 0.083000\n",
      "Largest haplotype segment: 96.045M\n",
      "Most-variants-phased (MVP) segment: 66402\n",
      "\n",
      "======================chr14=======================\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 51860\n",
      "Percentage of phased SNVs: 99.877\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.877\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.197000\n",
      "Blockwise Hamming distance: 57\n",
      "Blockwise Hamming rate: 0.110000\n",
      "Largest haplotype segment: 88.262M\n",
      "Most-variants-phased (MVP) segment: 51860\n",
      "\n",
      "======================chr15=======================\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 47386\n",
      "Percentage of phased SNVs: 98.791\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.791\n",
      "Number of switch errors: 490\n",
      "Switch error rate: 1.034000\n",
      "Blockwise Hamming distance: 572\n",
      "Blockwise Hamming rate: 1.207000\n",
      "Largest haplotype segment: 82.463M\n",
      "Most-variants-phased (MVP) segment: 47386\n",
      "\n",
      "======================chr16=======================\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 45313\n",
      "Percentage of phased SNVs: 98.954\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.954\n",
      "Number of switch errors: 508\n",
      "Switch error rate: 1.121000\n",
      "Blockwise Hamming distance: 435\n",
      "Blockwise Hamming rate: 0.960000\n",
      "Largest haplotype segment: 90.057M\n",
      "Most-variants-phased (MVP) segment: 45313\n",
      "\n",
      "======================chr17=======================\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 36626\n",
      "Percentage of phased SNVs: 99.425\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.424\n",
      "Number of switch errors: 192\n",
      "Switch error rate: 0.524000\n",
      "Blockwise Hamming distance: 151\n",
      "Blockwise Hamming rate: 0.412000\n",
      "Largest haplotype segment: 80.874M\n",
      "Most-variants-phased (MVP) segment: 36626\n",
      "\n",
      "======================chr18=======================\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 45694\n",
      "Percentage of phased SNVs: 99.897\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.897\n",
      "Number of switch errors: 80\n",
      "Switch error rate: 0.175000\n",
      "Blockwise Hamming distance: 40\n",
      "Blockwise Hamming rate: 0.088000\n",
      "Largest haplotype segment: 77.932M\n",
      "Most-variants-phased (MVP) segment: 45694\n",
      "\n",
      "======================chr19=======================\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 23932\n",
      "Percentage of phased SNVs: 98.291\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.291\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.426000\n",
      "Blockwise Hamming distance: 82\n",
      "Blockwise Hamming rate: 0.343000\n",
      "Largest haplotype segment: 58.827M\n",
      "Most-variants-phased (MVP) segment: 23932\n",
      "\n",
      "======================chr20=======================\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 32717\n",
      "Percentage of phased SNVs: 99.674\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.674\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.312000\n",
      "Blockwise Hamming distance: 76\n",
      "Blockwise Hamming rate: 0.232000\n",
      "Largest haplotype segment: 62.903M\n",
      "Most-variants-phased (MVP) segment: 32717\n",
      "\n",
      "======================chr21=======================\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 21890\n",
      "Percentage of phased SNVs: 98.285\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.285\n",
      "Number of switch errors: 301\n",
      "Switch error rate: 1.375000\n",
      "Blockwise Hamming distance: 236\n",
      "Blockwise Hamming rate: 1.078000\n",
      "Largest haplotype segment: 37.396M\n",
      "Most-variants-phased (MVP) segment: 21890\n",
      "\n",
      "======================chr22=======================\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 15260\n",
      "Percentage of phased SNVs: 99.478\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.478\n",
      "Number of switch errors: 38\n",
      "Switch error rate: 0.249000\n",
      "Blockwise Hamming distance: 24\n",
      "Blockwise Hamming rate: 0.157000\n",
      "Largest haplotype segment: 35.145M\n",
      "Most-variants-phased (MVP) segment: 15260\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[3].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[3]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = []\n",
    "    hapPos = []\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            \n",
    "            if line.startswith('hap1'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            hapBlock.append(lst[1])\n",
    "            hapPos.append(lst[0].split(':')[1])\n",
    "            #hapPos.append(lst[0])\n",
    "    \n",
    "    \n",
    "    numberOfSegments = 1\n",
    "    numberOfSNPs = len(hapBlock)\n",
    "    \n",
    "    hapCount = Counter(hapBlock)\n",
    "    \n",
    "    phasedSNPs = numberOfSNPs-hapCount['7']\n",
    "    completeness = round((phasedSNPs-1)/(numberOfSNPs-1)*100,3)\n",
    "    \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs*100,3)\n",
    "    \n",
    "    hammingDistance = min(hapCount['0'], hapCount['1'])\n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    LHSLength = round((int(hapPos[-1])-int(hapPos[0]))/1000000,3)\n",
    "    MVP = phasedSNPs\n",
    "    \n",
    "    hapClean = [hap for hap in hapBlock if hap != '7']\n",
    "    \n",
    "    v1 = hapClean[1:]\n",
    "    v2 = hapClean[:-1]\n",
    "        \n",
    "    vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "    switchError = sum(vd)\n",
    "    \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "     \n",
    "    metrics[chrom] = [numberOfSNPs, phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs)\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = open('whatshap/metrics.Hapi.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 120708\n",
      "Percentage of phased SNVs: 93.053\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 93.053\n",
      "Number of switch errors: 4623\n",
      "Switch error rate: 3.830000\n",
      "Blockwise Hamming distance: 6790\n",
      "Blockwise Hamming rate: 5.625000\n",
      "Largest haplotype segment: 248.393M\n",
      "Most-variants-phased (MVP) segment: 120708\n",
      "\n",
      "=======================chr2=======================\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 138930\n",
      "Percentage of phased SNVs: 98.067\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.067\n",
      "Number of switch errors: 1298\n",
      "Switch error rate: 0.934000\n",
      "Blockwise Hamming distance: 1260\n",
      "Blockwise Hamming rate: 0.907000\n",
      "Largest haplotype segment: 242.959M\n",
      "Most-variants-phased (MVP) segment: 138930\n",
      "\n",
      "=======================chr3=======================\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 127067\n",
      "Percentage of phased SNVs: 99.859\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.859\n",
      "Number of switch errors: 294\n",
      "Switch error rate: 0.231000\n",
      "Blockwise Hamming distance: 202\n",
      "Blockwise Hamming rate: 0.159000\n",
      "Largest haplotype segment: 197.794M\n",
      "Most-variants-phased (MVP) segment: 127067\n",
      "\n",
      "=======================chr4=======================\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 129980\n",
      "Percentage of phased SNVs: 99.868\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.868\n",
      "Number of switch errors: 438\n",
      "Switch error rate: 0.337000\n",
      "Blockwise Hamming distance: 546\n",
      "Blockwise Hamming rate: 0.420000\n",
      "Largest haplotype segment: 191.009M\n",
      "Most-variants-phased (MVP) segment: 129980\n",
      "\n",
      "=======================chr5=======================\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 114770\n",
      "Percentage of phased SNVs: 99.81\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.81\n",
      "Number of switch errors: 245\n",
      "Switch error rate: 0.213000\n",
      "Blockwise Hamming distance: 165\n",
      "Blockwise Hamming rate: 0.144000\n",
      "Largest haplotype segment: 180.652M\n",
      "Most-variants-phased (MVP) segment: 114770\n",
      "\n",
      "=======================chr6=======================\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 113218\n",
      "Percentage of phased SNVs: 99.45\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.45\n",
      "Number of switch errors: 642\n",
      "Switch error rate: 0.567000\n",
      "Blockwise Hamming distance: 445\n",
      "Blockwise Hamming rate: 0.393000\n",
      "Largest haplotype segment: 170.71M\n",
      "Most-variants-phased (MVP) segment: 113218\n",
      "\n",
      "=======================chr7=======================\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 97029\n",
      "Percentage of phased SNVs: 99.816\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.816\n",
      "Number of switch errors: 250\n",
      "Switch error rate: 0.258000\n",
      "Blockwise Hamming distance: 169\n",
      "Blockwise Hamming rate: 0.174000\n",
      "Largest haplotype segment: 159.066M\n",
      "Most-variants-phased (MVP) segment: 97029\n",
      "\n",
      "=======================chr8=======================\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 95561\n",
      "Percentage of phased SNVs: 99.947\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.947\n",
      "Number of switch errors: 82\n",
      "Switch error rate: 0.086000\n",
      "Blockwise Hamming distance: 46\n",
      "Blockwise Hamming rate: 0.048000\n",
      "Largest haplotype segment: 146.137M\n",
      "Most-variants-phased (MVP) segment: 95561\n",
      "\n",
      "=======================chr9=======================\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 69159\n",
      "Percentage of phased SNVs: 99.69\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.69\n",
      "Number of switch errors: 275\n",
      "Switch error rate: 0.398000\n",
      "Blockwise Hamming distance: 294\n",
      "Blockwise Hamming rate: 0.425000\n",
      "Largest haplotype segment: 140.973M\n",
      "Most-variants-phased (MVP) segment: 69159\n",
      "\n",
      "======================chr10=======================\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 88668\n",
      "Percentage of phased SNVs: 99.791\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.791\n",
      "Number of switch errors: 236\n",
      "Switch error rate: 0.266000\n",
      "Blockwise Hamming distance: 144\n",
      "Blockwise Hamming rate: 0.162000\n",
      "Largest haplotype segment: 135.371M\n",
      "Most-variants-phased (MVP) segment: 88668\n",
      "\n",
      "======================chr11=======================\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 86463\n",
      "Percentage of phased SNVs: 99.891\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.891\n",
      "Number of switch errors: 158\n",
      "Switch error rate: 0.183000\n",
      "Blockwise Hamming distance: 101\n",
      "Blockwise Hamming rate: 0.117000\n",
      "Largest haplotype segment: 134.713M\n",
      "Most-variants-phased (MVP) segment: 86463\n",
      "\n",
      "======================chr12=======================\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 77794\n",
      "Percentage of phased SNVs: 99.92\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.92\n",
      "Number of switch errors: 124\n",
      "Switch error rate: 0.159000\n",
      "Blockwise Hamming distance: 75\n",
      "Blockwise Hamming rate: 0.096000\n",
      "Largest haplotype segment: 133.619M\n",
      "Most-variants-phased (MVP) segment: 77794\n",
      "\n",
      "======================chr13=======================\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 66394\n",
      "Percentage of phased SNVs: 99.896\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.896\n",
      "Number of switch errors: 96\n",
      "Switch error rate: 0.145000\n",
      "Blockwise Hamming distance: 54\n",
      "Blockwise Hamming rate: 0.081000\n",
      "Largest haplotype segment: 96.045M\n",
      "Most-variants-phased (MVP) segment: 66394\n",
      "\n",
      "======================chr14=======================\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 51860\n",
      "Percentage of phased SNVs: 99.877\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.877\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.197000\n",
      "Blockwise Hamming distance: 57\n",
      "Blockwise Hamming rate: 0.110000\n",
      "Largest haplotype segment: 88.262M\n",
      "Most-variants-phased (MVP) segment: 51860\n",
      "\n",
      "======================chr15=======================\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 46688\n",
      "Percentage of phased SNVs: 97.336\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.336\n",
      "Number of switch errors: 776\n",
      "Switch error rate: 1.662000\n",
      "Blockwise Hamming distance: 689\n",
      "Blockwise Hamming rate: 1.476000\n",
      "Largest haplotype segment: 82.463M\n",
      "Most-variants-phased (MVP) segment: 46688\n",
      "\n",
      "======================chr16=======================\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 45282\n",
      "Percentage of phased SNVs: 98.886\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 98.886\n",
      "Number of switch errors: 520\n",
      "Switch error rate: 1.148000\n",
      "Blockwise Hamming distance: 443\n",
      "Blockwise Hamming rate: 0.978000\n",
      "Largest haplotype segment: 90.057M\n",
      "Most-variants-phased (MVP) segment: 45282\n",
      "\n",
      "======================chr17=======================\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 36656\n",
      "Percentage of phased SNVs: 99.506\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.506\n",
      "Number of switch errors: 192\n",
      "Switch error rate: 0.524000\n",
      "Blockwise Hamming distance: 151\n",
      "Blockwise Hamming rate: 0.412000\n",
      "Largest haplotype segment: 80.874M\n",
      "Most-variants-phased (MVP) segment: 36656\n",
      "\n",
      "======================chr18=======================\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 45695\n",
      "Percentage of phased SNVs: 99.899\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.899\n",
      "Number of switch errors: 80\n",
      "Switch error rate: 0.175000\n",
      "Blockwise Hamming distance: 40\n",
      "Blockwise Hamming rate: 0.088000\n",
      "Largest haplotype segment: 77.932M\n",
      "Most-variants-phased (MVP) segment: 45695\n",
      "\n",
      "======================chr19=======================\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 24259\n",
      "Percentage of phased SNVs: 99.634\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.634\n",
      "Number of switch errors: 118\n",
      "Switch error rate: 0.486000\n",
      "Blockwise Hamming distance: 102\n",
      "Blockwise Hamming rate: 0.420000\n",
      "Largest haplotype segment: 58.827M\n",
      "Most-variants-phased (MVP) segment: 24259\n",
      "\n",
      "======================chr20=======================\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 32684\n",
      "Percentage of phased SNVs: 99.573\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.573\n",
      "Number of switch errors: 110\n",
      "Switch error rate: 0.337000\n",
      "Blockwise Hamming distance: 179\n",
      "Blockwise Hamming rate: 0.548000\n",
      "Largest haplotype segment: 62.903M\n",
      "Most-variants-phased (MVP) segment: 32684\n",
      "\n",
      "======================chr21=======================\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 22087\n",
      "Percentage of phased SNVs: 99.169\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.169\n",
      "Number of switch errors: 277\n",
      "Switch error rate: 1.254000\n",
      "Blockwise Hamming distance: 172\n",
      "Blockwise Hamming rate: 0.779000\n",
      "Largest haplotype segment: 37.396M\n",
      "Most-variants-phased (MVP) segment: 22087\n",
      "\n",
      "======================chr22=======================\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 15323\n",
      "Percentage of phased SNVs: 99.889\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 99.889\n",
      "Number of switch errors: 48\n",
      "Switch error rate: 0.313000\n",
      "Blockwise Hamming distance: 31\n",
      "Blockwise Hamming rate: 0.202000\n",
      "Largest haplotype segment: 35.145M\n",
      "Most-variants-phased (MVP) segment: 15323\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('Comparison/GenomeResearch/Consistency/Cell2.sperm11.*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[2].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[2]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = []\n",
    "    hapPos = []\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            \n",
    "            if line.startswith('hap1'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            hapBlock.append(lst[1])\n",
    "            #hapPos.append(lst[0].split(':')[1])\n",
    "            hapPos.append(lst[0])\n",
    "    \n",
    "    \n",
    "    numberOfSegments = 1\n",
    "    numberOfSNPs = len(hapBlock)\n",
    "    \n",
    "    hapCount = Counter(hapBlock)\n",
    "    \n",
    "    phasedSNPs = numberOfSNPs-hapCount['7']\n",
    "    completeness = round((phasedSNPs-1)/(numberOfSNPs-1)*100,3)\n",
    "    \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs*100,3)\n",
    "    \n",
    "    hammingDistance = min(hapCount['0'], hapCount['1'])\n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    LHSLength = round((int(hapPos[-1])-int(hapPos[0]))/1000000,3)\n",
    "    MVP = phasedSNPs\n",
    "    \n",
    "    hapClean = [hap for hap in hapBlock if hap != '7']\n",
    "    \n",
    "    v1 = hapClean[1:]\n",
    "    v2 = hapClean[:-1]\n",
    "        \n",
    "    vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "    switchError = sum(vd)\n",
    "    \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "     \n",
    "    metrics[chrom] = [numberOfSNPs, phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs)\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = open('whatshap/metrics.PHMM.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "88770663\n",
      "195037001\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 76142\n",
      "Percentage of phased SNVs: 58.697\n",
      "Number of phased blocks: 23095\n",
      "Completeness (NC): 40.893\n",
      "Number of switch errors: 616\n",
      "Switch error rate: 1.161000\n",
      "Blockwise Hamming distance: 631\n",
      "Blockwise Hamming rate: 0.829000\n",
      "Largest haplotype segment: 98.86M\n",
      "Most-variants-phased (MVP) segment: 191\n",
      "\n",
      "=======================chr2=======================\n",
      "62307899\n",
      "35870947\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 83960\n",
      "Percentage of phased SNVs: 59.265\n",
      "Number of phased blocks: 25571\n",
      "Completeness (NC): 41.215\n",
      "Number of switch errors: 489\n",
      "Switch error rate: 0.837000\n",
      "Blockwise Hamming distance: 595\n",
      "Blockwise Hamming rate: 0.709000\n",
      "Largest haplotype segment: 79.708M\n",
      "Most-variants-phased (MVP) segment: 308\n",
      "\n",
      "=======================chr3=======================\n",
      "50847250\n",
      "623978\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 77365\n",
      "Percentage of phased SNVs: 60.799\n",
      "Number of phased blocks: 22754\n",
      "Completeness (NC): 42.917\n",
      "Number of switch errors: 303\n",
      "Switch error rate: 0.555000\n",
      "Blockwise Hamming distance: 357\n",
      "Blockwise Hamming rate: 0.461000\n",
      "Largest haplotype segment: 81.711M\n",
      "Most-variants-phased (MVP) segment: 287\n",
      "\n",
      "=======================chr4=======================\n",
      "48806762\n",
      "181536946\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 83039\n",
      "Percentage of phased SNVs: 63.802\n",
      "Number of phased blocks: 23322\n",
      "Completeness (NC): 45.883\n",
      "Number of switch errors: 453\n",
      "Switch error rate: 0.759000\n",
      "Blockwise Hamming distance: 774\n",
      "Blockwise Hamming rate: 0.932000\n",
      "Largest haplotype segment: 58.192M\n",
      "Most-variants-phased (MVP) segment: 1554\n",
      "\n",
      "=======================chr5=======================\n",
      "91618448\n",
      "21523395\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 69544\n",
      "Percentage of phased SNVs: 60.479\n",
      "Number of phased blocks: 20757\n",
      "Completeness (NC): 42.428\n",
      "Number of switch errors: 296\n",
      "Switch error rate: 0.607000\n",
      "Blockwise Hamming distance: 433\n",
      "Blockwise Hamming rate: 0.623000\n",
      "Largest haplotype segment: 68.575M\n",
      "Most-variants-phased (MVP) segment: 539\n",
      "\n",
      "=======================chr6=======================\n",
      "21014556\n",
      "21014556\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 70672\n",
      "Percentage of phased SNVs: 62.078\n",
      "Number of phased blocks: 19466\n",
      "Completeness (NC): 44.979\n",
      "Number of switch errors: 964\n",
      "Switch error rate: 1.883000\n",
      "Blockwise Hamming distance: 1879\n",
      "Blockwise Hamming rate: 2.659000\n",
      "Largest haplotype segment: 87.992M\n",
      "Most-variants-phased (MVP) segment: 3683\n",
      "\n",
      "=======================chr7=======================\n",
      "40352072\n",
      "152103712\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 59112\n",
      "Percentage of phased SNVs: 60.81\n",
      "Number of phased blocks: 17643\n",
      "Completeness (NC): 42.66\n",
      "Number of switch errors: 255\n",
      "Switch error rate: 0.615000\n",
      "Blockwise Hamming distance: 322\n",
      "Blockwise Hamming rate: 0.545000\n",
      "Largest haplotype segment: 88.848M\n",
      "Most-variants-phased (MVP) segment: 111\n",
      "\n",
      "=======================chr8=======================\n",
      "20843176\n",
      "3218427\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 60196\n",
      "Percentage of phased SNVs: 62.959\n",
      "Number of phased blocks: 17349\n",
      "Completeness (NC): 44.813\n",
      "Number of switch errors: 141\n",
      "Switch error rate: 0.329000\n",
      "Blockwise Hamming distance: 136\n",
      "Blockwise Hamming rate: 0.226000\n",
      "Largest haplotype segment: 63.678M\n",
      "Most-variants-phased (MVP) segment: 175\n",
      "\n",
      "=======================chr9=======================\n",
      "17327271\n",
      "28069562\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 42081\n",
      "Percentage of phased SNVs: 60.658\n",
      "Number of phased blocks: 12562\n",
      "Completeness (NC): 42.551\n",
      "Number of switch errors: 193\n",
      "Switch error rate: 0.654000\n",
      "Blockwise Hamming distance: 264\n",
      "Blockwise Hamming rate: 0.627000\n",
      "Largest haplotype segment: 79.522M\n",
      "Most-variants-phased (MVP) segment: 261\n",
      "\n",
      "======================chr10=======================\n",
      "18282817\n",
      "5160613\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 54560\n",
      "Percentage of phased SNVs: 61.404\n",
      "Number of phased blocks: 16106\n",
      "Completeness (NC): 43.278\n",
      "Number of switch errors: 248\n",
      "Switch error rate: 0.645000\n",
      "Blockwise Hamming distance: 264\n",
      "Blockwise Hamming rate: 0.484000\n",
      "Largest haplotype segment: 68.869M\n",
      "Most-variants-phased (MVP) segment: 108\n",
      "\n",
      "======================chr11=======================\n",
      "29967308\n",
      "80605259\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 53405\n",
      "Percentage of phased SNVs: 61.699\n",
      "Number of phased blocks: 15506\n",
      "Completeness (NC): 43.785\n",
      "Number of switch errors: 174\n",
      "Switch error rate: 0.459000\n",
      "Blockwise Hamming distance: 273\n",
      "Blockwise Hamming rate: 0.511000\n",
      "Largest haplotype segment: 58.506M\n",
      "Most-variants-phased (MVP) segment: 127\n",
      "\n",
      "======================chr12=======================\n",
      "17555326\n",
      "20991130\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 45534\n",
      "Percentage of phased SNVs: 58.485\n",
      "Number of phased blocks: 14189\n",
      "Completeness (NC): 40.26\n",
      "Number of switch errors: 107\n",
      "Switch error rate: 0.341000\n",
      "Blockwise Hamming distance: 147\n",
      "Blockwise Hamming rate: 0.323000\n",
      "Largest haplotype segment: 92.293M\n",
      "Most-variants-phased (MVP) segment: 141\n",
      "\n",
      "======================chr13=======================\n",
      "29725658\n",
      "63604733\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 41189\n",
      "Percentage of phased SNVs: 61.973\n",
      "Number of phased blocks: 12125\n",
      "Completeness (NC): 43.73\n",
      "Number of switch errors: 122\n",
      "Switch error rate: 0.420000\n",
      "Blockwise Hamming distance: 198\n",
      "Blockwise Hamming rate: 0.481000\n",
      "Largest haplotype segment: 52.823M\n",
      "Most-variants-phased (MVP) segment: 571\n",
      "\n",
      "======================chr14=======================\n",
      "25682182\n",
      "39133778\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 30749\n",
      "Percentage of phased SNVs: 59.219\n",
      "Number of phased blocks: 9396\n",
      "Completeness (NC): 41.124\n",
      "Number of switch errors: 114\n",
      "Switch error rate: 0.534000\n",
      "Blockwise Hamming distance: 137\n",
      "Blockwise Hamming rate: 0.446000\n",
      "Largest haplotype segment: 57.582M\n",
      "Most-variants-phased (MVP) segment: 111\n",
      "\n",
      "======================chr15=======================\n",
      "24027283\n",
      "54342034\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 28117\n",
      "Percentage of phased SNVs: 58.619\n",
      "Number of phased blocks: 8747\n",
      "Completeness (NC): 40.383\n",
      "Number of switch errors: 276\n",
      "Switch error rate: 1.425000\n",
      "Blockwise Hamming distance: 310\n",
      "Blockwise Hamming rate: 1.103000\n",
      "Largest haplotype segment: 75.725M\n",
      "Most-variants-phased (MVP) segment: 158\n",
      "\n",
      "======================chr16=======================\n",
      "7023832\n",
      "5730814\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 27893\n",
      "Percentage of phased SNVs: 60.912\n",
      "Number of phased blocks: 8111\n",
      "Completeness (NC): 43.2\n",
      "Number of switch errors: 408\n",
      "Switch error rate: 2.062000\n",
      "Blockwise Hamming distance: 384\n",
      "Blockwise Hamming rate: 1.377000\n",
      "Largest haplotype segment: 71.203M\n",
      "Most-variants-phased (MVP) segment: 187\n",
      "\n",
      "======================chr17=======================\n",
      "3309852\n",
      "3309852\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 19593\n",
      "Percentage of phased SNVs: 53.187\n",
      "Number of phased blocks: 6221\n",
      "Completeness (NC): 36.299\n",
      "Number of switch errors: 228\n",
      "Switch error rate: 1.705000\n",
      "Blockwise Hamming distance: 245\n",
      "Blockwise Hamming rate: 1.250000\n",
      "Largest haplotype segment: 60.721M\n",
      "Most-variants-phased (MVP) segment: 231\n",
      "\n",
      "======================chr18=======================\n",
      "1386741\n",
      "32151047\n",
      "69993799\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 26651\n",
      "Percentage of phased SNVs: 58.265\n",
      "Number of phased blocks: 8407\n",
      "Completeness (NC): 39.885\n",
      "Number of switch errors: 74\n",
      "Switch error rate: 0.406000\n",
      "Blockwise Hamming distance: 83\n",
      "Blockwise Hamming rate: 0.311000\n",
      "Largest haplotype segment: 67.326M\n",
      "Most-variants-phased (MVP) segment: 79\n",
      "\n",
      "======================chr19=======================\n",
      "32934755\n",
      "12519638\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 13416\n",
      "Percentage of phased SNVs: 55.101\n",
      "Number of phased blocks: 4143\n",
      "Completeness (NC): 38.085\n",
      "Number of switch errors: 41\n",
      "Switch error rate: 0.442000\n",
      "Blockwise Hamming distance: 42\n",
      "Blockwise Hamming rate: 0.313000\n",
      "Largest haplotype segment: 0.488M\n",
      "Most-variants-phased (MVP) segment: 147\n",
      "\n",
      "======================chr20=======================\n",
      "9926657\n",
      "6857854\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 19000\n",
      "Percentage of phased SNVs: 57.884\n",
      "Number of phased blocks: 5919\n",
      "Completeness (NC): 39.852\n",
      "Number of switch errors: 119\n",
      "Switch error rate: 0.910000\n",
      "Blockwise Hamming distance: 151\n",
      "Blockwise Hamming rate: 0.795000\n",
      "Largest haplotype segment: 9.493M\n",
      "Most-variants-phased (MVP) segment: 110\n",
      "\n",
      "======================chr21=======================\n",
      "19940635\n",
      "11037148\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 14003\n",
      "Percentage of phased SNVs: 62.873\n",
      "Number of phased blocks: 3938\n",
      "Completeness (NC): 45.191\n",
      "Number of switch errors: 304\n",
      "Switch error rate: 3.020000\n",
      "Blockwise Hamming distance: 317\n",
      "Blockwise Hamming rate: 2.264000\n",
      "Largest haplotype segment: 23.899M\n",
      "Most-variants-phased (MVP) segment: 159\n",
      "\n",
      "======================chr22=======================\n",
      "31665909\n",
      "39293026\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 7888\n",
      "Percentage of phased SNVs: 51.421\n",
      "Number of phased blocks: 2597\n",
      "Completeness (NC): 34.492\n",
      "Number of switch errors: 16\n",
      "Switch error rate: 0.302000\n",
      "Blockwise Hamming distance: 13\n",
      "Blockwise Hamming rate: 0.165000\n",
      "Largest haplotype segment: 4.235M\n",
      "Most-variants-phased (MVP) segment: 51\n",
      "\n",
      "=======================chr1=======================\n",
      "111579512\n",
      "111579512\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 73124\n",
      "Percentage of phased SNVs: 56.371\n",
      "Number of phased blocks: 23510\n",
      "Completeness (NC): 38.247\n",
      "Number of switch errors: 440\n",
      "Switch error rate: 0.887000\n",
      "Blockwise Hamming distance: 408\n",
      "Blockwise Hamming rate: 0.558000\n",
      "Largest haplotype segment: 0.008M\n",
      "Most-variants-phased (MVP) segment: 55\n",
      "\n",
      "=======================chr2=======================\n",
      "133067542\n",
      "133047563\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 80570\n",
      "Percentage of phased SNVs: 56.872\n",
      "Number of phased blocks: 26033\n",
      "Completeness (NC): 38.496\n",
      "Number of switch errors: 343\n",
      "Switch error rate: 0.629000\n",
      "Blockwise Hamming distance: 371\n",
      "Blockwise Hamming rate: 0.460000\n",
      "Largest haplotype segment: 0.013M\n",
      "Most-variants-phased (MVP) segment: 123\n",
      "\n",
      "=======================chr3=======================\n",
      "631759\n",
      "631759\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 74431\n",
      "Percentage of phased SNVs: 58.493\n",
      "Number of phased blocks: 23266\n",
      "Completeness (NC): 40.209\n",
      "Number of switch errors: 184\n",
      "Switch error rate: 0.360000\n",
      "Blockwise Hamming distance: 181\n",
      "Blockwise Hamming rate: 0.243000\n",
      "Largest haplotype segment: 0.018M\n",
      "Most-variants-phased (MVP) segment: 166\n",
      "\n",
      "=======================chr4=======================\n",
      "190622165\n",
      "190622165\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 80042\n",
      "Percentage of phased SNVs: 61.499\n",
      "Number of phased blocks: 24045\n",
      "Completeness (NC): 43.024\n",
      "Number of switch errors: 330\n",
      "Switch error rate: 0.589000\n",
      "Blockwise Hamming distance: 385\n",
      "Blockwise Hamming rate: 0.481000\n",
      "Largest haplotype segment: 0.034M\n",
      "Most-variants-phased (MVP) segment: 502\n",
      "\n",
      "=======================chr5=======================\n",
      "21560130\n",
      "21560130\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 66790\n",
      "Percentage of phased SNVs: 58.084\n",
      "Number of phased blocks: 21201\n",
      "Completeness (NC): 39.647\n",
      "Number of switch errors: 205\n",
      "Switch error rate: 0.450000\n",
      "Blockwise Hamming distance: 233\n",
      "Blockwise Hamming rate: 0.349000\n",
      "Largest haplotype segment: 0.012M\n",
      "Most-variants-phased (MVP) segment: 163\n",
      "\n",
      "=======================chr6=======================\n",
      "57361278\n",
      "57361278\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 68273\n",
      "Percentage of phased SNVs: 59.971\n",
      "Number of phased blocks: 19984\n",
      "Completeness (NC): 42.417\n",
      "Number of switch errors: 805\n",
      "Switch error rate: 1.667000\n",
      "Blockwise Hamming distance: 1002\n",
      "Blockwise Hamming rate: 1.468000\n",
      "Largest haplotype segment: 0.04M\n",
      "Most-variants-phased (MVP) segment: 396\n",
      "\n",
      "=======================chr7=======================\n",
      "10088010\n",
      "152103712\n",
      "152103712\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 56824\n",
      "Percentage of phased SNVs: 58.456\n",
      "Number of phased blocks: 17936\n",
      "Completeness (NC): 40.005\n",
      "Number of switch errors: 156\n",
      "Switch error rate: 0.401000\n",
      "Blockwise Hamming distance: 159\n",
      "Blockwise Hamming rate: 0.280000\n",
      "Largest haplotype segment: 0.01M\n",
      "Most-variants-phased (MVP) segment: 111\n",
      "\n",
      "=======================chr8=======================\n",
      "11231249\n",
      "3996983\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 57995\n",
      "Percentage of phased SNVs: 60.657\n",
      "Number of phased blocks: 17792\n",
      "Completeness (NC): 42.048\n",
      "Number of switch errors: 78\n",
      "Switch error rate: 0.194000\n",
      "Blockwise Hamming distance: 72\n",
      "Blockwise Hamming rate: 0.124000\n",
      "Largest haplotype segment: 0.009M\n",
      "Most-variants-phased (MVP) segment: 83\n",
      "\n",
      "=======================chr9=======================\n",
      "9381760\n",
      "9390765\n",
      "32060229\n",
      "68412384\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 40482\n",
      "Percentage of phased SNVs: 58.353\n",
      "Number of phased blocks: 12790\n",
      "Completeness (NC): 39.917\n",
      "Number of switch errors: 144\n",
      "Switch error rate: 0.520000\n",
      "Blockwise Hamming distance: 149\n",
      "Blockwise Hamming rate: 0.368000\n",
      "Largest haplotype segment: 0.007M\n",
      "Most-variants-phased (MVP) segment: 51\n",
      "\n",
      "======================chr10=======================\n",
      "5168978\n",
      "55319395\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 52347\n",
      "Percentage of phased SNVs: 58.913\n",
      "Number of phased blocks: 16368\n",
      "Completeness (NC): 40.492\n",
      "Number of switch errors: 165\n",
      "Switch error rate: 0.459000\n",
      "Blockwise Hamming distance: 154\n",
      "Blockwise Hamming rate: 0.294000\n",
      "Largest haplotype segment: 0.01M\n",
      "Most-variants-phased (MVP) segment: 73\n",
      "\n",
      "======================chr11=======================\n",
      "80609285\n",
      "80609285\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 51427\n",
      "Percentage of phased SNVs: 59.414\n",
      "Number of phased blocks: 15895\n",
      "Completeness (NC): 41.05\n",
      "Number of switch errors: 81\n",
      "Switch error rate: 0.228000\n",
      "Blockwise Hamming distance: 108\n",
      "Blockwise Hamming rate: 0.210000\n",
      "Largest haplotype segment: 0.011M\n",
      "Most-variants-phased (MVP) segment: 72\n",
      "\n",
      "======================chr12=======================\n",
      "126554565\n",
      "17879598\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 43803\n",
      "Percentage of phased SNVs: 56.262\n",
      "Number of phased blocks: 14353\n",
      "Completeness (NC): 37.826\n",
      "Number of switch errors: 50\n",
      "Switch error rate: 0.170000\n",
      "Blockwise Hamming distance: 63\n",
      "Blockwise Hamming rate: 0.144000\n",
      "Largest haplotype segment: 0.009M\n",
      "Most-variants-phased (MVP) segment: 59\n",
      "\n",
      "======================chr13=======================\n",
      "63605744\n",
      "63605744\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 39653\n",
      "Percentage of phased SNVs: 59.662\n",
      "Number of phased blocks: 12387\n",
      "Completeness (NC): 41.024\n",
      "Number of switch errors: 80\n",
      "Switch error rate: 0.293000\n",
      "Blockwise Hamming distance: 90\n",
      "Blockwise Hamming rate: 0.227000\n",
      "Largest haplotype segment: 0.018M\n",
      "Most-variants-phased (MVP) segment: 284\n",
      "\n",
      "======================chr14=======================\n",
      "34714162\n",
      "96078162\n",
      "34714162\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 29485\n",
      "Percentage of phased SNVs: 56.785\n",
      "Number of phased blocks: 9570\n",
      "Completeness (NC): 38.354\n",
      "Number of switch errors: 58\n",
      "Switch error rate: 0.291000\n",
      "Blockwise Hamming distance: 80\n",
      "Blockwise Hamming rate: 0.271000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 69\n",
      "\n",
      "======================chr15=======================\n",
      "22402142\n",
      "55135408\n",
      "55152117\n",
      "100388486\n",
      "55140913\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 26914\n",
      "Percentage of phased SNVs: 56.111\n",
      "Number of phased blocks: 8791\n",
      "Completeness (NC): 37.783\n",
      "Number of switch errors: 208\n",
      "Switch error rate: 1.148000\n",
      "Blockwise Hamming distance: 201\n",
      "Blockwise Hamming rate: 0.747000\n",
      "Largest haplotype segment: 0.005M\n",
      "Most-variants-phased (MVP) segment: 46\n",
      "\n",
      "======================chr16=======================\n",
      "76441464\n",
      "76441464\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 26837\n",
      "Percentage of phased SNVs: 58.606\n",
      "Number of phased blocks: 8160\n",
      "Completeness (NC): 40.787\n",
      "Number of switch errors: 353\n",
      "Switch error rate: 1.890000\n",
      "Blockwise Hamming distance: 292\n",
      "Blockwise Hamming rate: 1.088000\n",
      "Largest haplotype segment: 0.01M\n",
      "Most-variants-phased (MVP) segment: 83\n",
      "\n",
      "======================chr17=======================\n",
      "39134098\n",
      "63991674\n",
      "39134098\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 18549\n",
      "Percentage of phased SNVs: 50.353\n",
      "Number of phased blocks: 6137\n",
      "Completeness (NC): 33.693\n",
      "Number of switch errors: 173\n",
      "Switch error rate: 1.394000\n",
      "Blockwise Hamming distance: 168\n",
      "Blockwise Hamming rate: 0.906000\n",
      "Largest haplotype segment: 0.006M\n",
      "Most-variants-phased (MVP) segment: 44\n",
      "\n",
      "======================chr18=======================\n",
      "2295048\n",
      "32162754\n",
      "70000592\n",
      "70000592\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 25539\n",
      "Percentage of phased SNVs: 55.834\n",
      "Number of phased blocks: 8429\n",
      "Completeness (NC): 37.406\n",
      "Number of switch errors: 31\n",
      "Switch error rate: 0.181000\n",
      "Blockwise Hamming distance: 31\n",
      "Blockwise Hamming rate: 0.121000\n",
      "Largest haplotype segment: 0.005M\n",
      "Most-variants-phased (MVP) segment: 49\n",
      "\n",
      "======================chr19=======================\n",
      "12543251\n",
      "12543251\n",
      "20751225\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 12828\n",
      "Percentage of phased SNVs: 52.686\n",
      "Number of phased blocks: 4094\n",
      "Completeness (NC): 35.872\n",
      "Number of switch errors: 14\n",
      "Switch error rate: 0.160000\n",
      "Blockwise Hamming distance: 12\n",
      "Blockwise Hamming rate: 0.094000\n",
      "Largest haplotype segment: 0.008M\n",
      "Most-variants-phased (MVP) segment: 45\n",
      "\n",
      "======================chr20=======================\n",
      "26200957\n",
      "26200957\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 18184\n",
      "Percentage of phased SNVs: 55.398\n",
      "Number of phased blocks: 5972\n",
      "Completeness (NC): 37.204\n",
      "Number of switch errors: 89\n",
      "Switch error rate: 0.729000\n",
      "Blockwise Hamming distance: 112\n",
      "Blockwise Hamming rate: 0.616000\n",
      "Largest haplotype segment: 0.01M\n",
      "Most-variants-phased (MVP) segment: 82\n",
      "\n",
      "======================chr21=======================\n",
      "11042288\n",
      "11042288\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 13517\n",
      "Percentage of phased SNVs: 60.691\n",
      "Number of phased blocks: 4021\n",
      "Completeness (NC): 42.636\n",
      "Number of switch errors: 272\n",
      "Switch error rate: 2.864000\n",
      "Blockwise Hamming distance: 269\n",
      "Blockwise Hamming rate: 1.990000\n",
      "Largest haplotype segment: 0.016M\n",
      "Most-variants-phased (MVP) segment: 125\n",
      "\n",
      "======================chr22=======================\n",
      "22712261\n",
      "35331402\n",
      "35337607\n",
      "35469757\n",
      "39289802\n",
      "39299216\n",
      "39299216\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 7463\n",
      "Percentage of phased SNVs: 48.651\n",
      "Number of phased blocks: 2528\n",
      "Completeness (NC): 32.171\n",
      "Number of switch errors: 11\n",
      "Switch error rate: 0.223000\n",
      "Blockwise Hamming distance: 10\n",
      "Blockwise Hamming rate: 0.134000\n",
      "Largest haplotype segment: 0.003M\n",
      "Most-variants-phased (MVP) segment: 39\n",
      "\n",
      "=======================chr1=======================\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 125982\n",
      "Percentage of phased SNVs: 97.118\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.118\n",
      "Number of switch errors: 537\n",
      "Switch error rate: 0.426000\n",
      "Blockwise Hamming distance: 344\n",
      "Blockwise Hamming rate: 0.273000\n",
      "Largest haplotype segment: 248.393M\n",
      "Most-variants-phased (MVP) segment: 125982\n",
      "\n",
      "=======================chr2=======================\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 137760\n",
      "Percentage of phased SNVs: 97.241\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.241\n",
      "Number of switch errors: 448\n",
      "Switch error rate: 0.325000\n",
      "Blockwise Hamming distance: 318\n",
      "Blockwise Hamming rate: 0.231000\n",
      "Largest haplotype segment: 242.959M\n",
      "Most-variants-phased (MVP) segment: 137760\n",
      "\n",
      "=======================chr3=======================\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 124396\n",
      "Percentage of phased SNVs: 97.759\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.759\n",
      "Number of switch errors: 222\n",
      "Switch error rate: 0.178000\n",
      "Blockwise Hamming distance: 145\n",
      "Blockwise Hamming rate: 0.117000\n",
      "Largest haplotype segment: 197.794M\n",
      "Most-variants-phased (MVP) segment: 124396\n",
      "\n",
      "=======================chr4=======================\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 127546\n",
      "Percentage of phased SNVs: 97.998\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.998\n",
      "Number of switch errors: 434\n",
      "Switch error rate: 0.340000\n",
      "Blockwise Hamming distance: 524\n",
      "Blockwise Hamming rate: 0.411000\n",
      "Largest haplotype segment: 191.009M\n",
      "Most-variants-phased (MVP) segment: 127546\n",
      "\n",
      "=======================chr5=======================\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 112405\n",
      "Percentage of phased SNVs: 97.754\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.754\n",
      "Number of switch errors: 241\n",
      "Switch error rate: 0.214000\n",
      "Blockwise Hamming distance: 165\n",
      "Blockwise Hamming rate: 0.147000\n",
      "Largest haplotype segment: 180.471M\n",
      "Most-variants-phased (MVP) segment: 112405\n",
      "\n",
      "=======================chr6=======================\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 110132\n",
      "Percentage of phased SNVs: 96.739\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.739\n",
      "Number of switch errors: 648\n",
      "Switch error rate: 0.588000\n",
      "Blockwise Hamming distance: 457\n",
      "Blockwise Hamming rate: 0.415000\n",
      "Largest haplotype segment: 170.71M\n",
      "Most-variants-phased (MVP) segment: 110132\n",
      "\n",
      "=======================chr7=======================\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 94132\n",
      "Percentage of phased SNVs: 96.836\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.836\n",
      "Number of switch errors: 232\n",
      "Switch error rate: 0.246000\n",
      "Blockwise Hamming distance: 139\n",
      "Blockwise Hamming rate: 0.148000\n",
      "Largest haplotype segment: 159.056M\n",
      "Most-variants-phased (MVP) segment: 94132\n",
      "\n",
      "=======================chr8=======================\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 93216\n",
      "Percentage of phased SNVs: 97.494\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.494\n",
      "Number of switch errors: 86\n",
      "Switch error rate: 0.092000\n",
      "Blockwise Hamming distance: 45\n",
      "Blockwise Hamming rate: 0.048000\n",
      "Largest haplotype segment: 146.137M\n",
      "Most-variants-phased (MVP) segment: 93216\n",
      "\n",
      "=======================chr9=======================\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 67272\n",
      "Percentage of phased SNVs: 96.97\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.97\n",
      "Number of switch errors: 265\n",
      "Switch error rate: 0.394000\n",
      "Blockwise Hamming distance: 224\n",
      "Blockwise Hamming rate: 0.333000\n",
      "Largest haplotype segment: 140.973M\n",
      "Most-variants-phased (MVP) segment: 67272\n",
      "\n",
      "======================chr10=======================\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 86139\n",
      "Percentage of phased SNVs: 96.944\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.944\n",
      "Number of switch errors: 260\n",
      "Switch error rate: 0.302000\n",
      "Blockwise Hamming distance: 201\n",
      "Blockwise Hamming rate: 0.233000\n",
      "Largest haplotype segment: 135.371M\n",
      "Most-variants-phased (MVP) segment: 86139\n",
      "\n",
      "======================chr11=======================\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 84186\n",
      "Percentage of phased SNVs: 97.261\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.261\n",
      "Number of switch errors: 154\n",
      "Switch error rate: 0.183000\n",
      "Blockwise Hamming distance: 100\n",
      "Blockwise Hamming rate: 0.119000\n",
      "Largest haplotype segment: 134.713M\n",
      "Most-variants-phased (MVP) segment: 84186\n",
      "\n",
      "======================chr12=======================\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 75618\n",
      "Percentage of phased SNVs: 97.125\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.125\n",
      "Number of switch errors: 110\n",
      "Switch error rate: 0.145000\n",
      "Blockwise Hamming distance: 72\n",
      "Blockwise Hamming rate: 0.095000\n",
      "Largest haplotype segment: 133.619M\n",
      "Most-variants-phased (MVP) segment: 75618\n",
      "\n",
      "======================chr13=======================\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 64907\n",
      "Percentage of phased SNVs: 97.659\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.659\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.157000\n",
      "Blockwise Hamming distance: 59\n",
      "Blockwise Hamming rate: 0.091000\n",
      "Largest haplotype segment: 96.045M\n",
      "Most-variants-phased (MVP) segment: 64907\n",
      "\n",
      "======================chr14=======================\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 50599\n",
      "Percentage of phased SNVs: 97.448\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.448\n",
      "Number of switch errors: 112\n",
      "Switch error rate: 0.221000\n",
      "Blockwise Hamming distance: 65\n",
      "Blockwise Hamming rate: 0.128000\n",
      "Largest haplotype segment: 88.262M\n",
      "Most-variants-phased (MVP) segment: 50599\n",
      "\n",
      "======================chr15=======================\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 45883\n",
      "Percentage of phased SNVs: 95.657\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 95.657\n",
      "Number of switch errors: 528\n",
      "Switch error rate: 1.151000\n",
      "Blockwise Hamming distance: 612\n",
      "Blockwise Hamming rate: 1.334000\n",
      "Largest haplotype segment: 82.463M\n",
      "Most-variants-phased (MVP) segment: 45883\n",
      "\n",
      "======================chr16=======================\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 43626\n",
      "Percentage of phased SNVs: 95.27\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 95.27\n",
      "Number of switch errors: 516\n",
      "Switch error rate: 1.183000\n",
      "Blockwise Hamming distance: 427\n",
      "Blockwise Hamming rate: 0.979000\n",
      "Largest haplotype segment: 90.052M\n",
      "Most-variants-phased (MVP) segment: 43626\n",
      "\n",
      "======================chr17=======================\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 34898\n",
      "Percentage of phased SNVs: 94.734\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.734\n",
      "Number of switch errors: 210\n",
      "Switch error rate: 0.602000\n",
      "Blockwise Hamming distance: 172\n",
      "Blockwise Hamming rate: 0.493000\n",
      "Largest haplotype segment: 80.874M\n",
      "Most-variants-phased (MVP) segment: 34898\n",
      "\n",
      "======================chr18=======================\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 44429\n",
      "Percentage of phased SNVs: 97.132\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.132\n",
      "Number of switch errors: 84\n",
      "Switch error rate: 0.189000\n",
      "Blockwise Hamming distance: 44\n",
      "Blockwise Hamming rate: 0.099000\n",
      "Largest haplotype segment: 77.932M\n",
      "Most-variants-phased (MVP) segment: 44429\n",
      "\n",
      "======================chr19=======================\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 23034\n",
      "Percentage of phased SNVs: 94.603\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.603\n",
      "Number of switch errors: 106\n",
      "Switch error rate: 0.460000\n",
      "Blockwise Hamming distance: 86\n",
      "Blockwise Hamming rate: 0.373000\n",
      "Largest haplotype segment: 58.827M\n",
      "Most-variants-phased (MVP) segment: 23034\n",
      "\n",
      "======================chr20=======================\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 31577\n",
      "Percentage of phased SNVs: 96.201\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.201\n",
      "Number of switch errors: 104\n",
      "Switch error rate: 0.329000\n",
      "Blockwise Hamming distance: 77\n",
      "Blockwise Hamming rate: 0.244000\n",
      "Largest haplotype segment: 62.903M\n",
      "Most-variants-phased (MVP) segment: 31577\n",
      "\n",
      "======================chr21=======================\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 21577\n",
      "Percentage of phased SNVs: 96.879\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.879\n",
      "Number of switch errors: 390\n",
      "Switch error rate: 1.808000\n",
      "Blockwise Hamming distance: 465\n",
      "Blockwise Hamming rate: 2.155000\n",
      "Largest haplotype segment: 37.396M\n",
      "Most-variants-phased (MVP) segment: 21577\n",
      "\n",
      "======================chr22=======================\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 14519\n",
      "Percentage of phased SNVs: 94.648\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.648\n",
      "Number of switch errors: 44\n",
      "Switch error rate: 0.303000\n",
      "Blockwise Hamming distance: 26\n",
      "Blockwise Hamming rate: 0.179000\n",
      "Largest haplotype segment: 35.145M\n",
      "Most-variants-phased (MVP) segment: 14519\n",
      "\n",
      "=======================chr1=======================\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 123548\n",
      "Percentage of phased SNVs: 95.242\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 95.242\n",
      "Number of switch errors: 1735\n",
      "Switch error rate: 1.404000\n",
      "Blockwise Hamming distance: 1772\n",
      "Blockwise Hamming rate: 1.434000\n",
      "Largest haplotype segment: 248.393M\n",
      "Most-variants-phased (MVP) segment: 123548\n",
      "\n",
      "=======================chr2=======================\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 134936\n",
      "Percentage of phased SNVs: 95.247\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 95.247\n",
      "Number of switch errors: 1468\n",
      "Switch error rate: 1.088000\n",
      "Blockwise Hamming distance: 1560\n",
      "Blockwise Hamming rate: 1.156000\n",
      "Largest haplotype segment: 242.959M\n",
      "Most-variants-phased (MVP) segment: 134936\n",
      "\n",
      "=======================chr3=======================\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 124387\n",
      "Percentage of phased SNVs: 97.752\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.752\n",
      "Number of switch errors: 228\n",
      "Switch error rate: 0.183000\n",
      "Blockwise Hamming distance: 151\n",
      "Blockwise Hamming rate: 0.121000\n",
      "Largest haplotype segment: 197.794M\n",
      "Most-variants-phased (MVP) segment: 124387\n",
      "\n",
      "=======================chr4=======================\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 126723\n",
      "Percentage of phased SNVs: 97.365\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.365\n",
      "Number of switch errors: 696\n",
      "Switch error rate: 0.549000\n",
      "Blockwise Hamming distance: 870\n",
      "Blockwise Hamming rate: 0.687000\n",
      "Largest haplotype segment: 191.009M\n",
      "Most-variants-phased (MVP) segment: 126723\n",
      "\n",
      "=======================chr5=======================\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 112405\n",
      "Percentage of phased SNVs: 97.754\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.754\n",
      "Number of switch errors: 241\n",
      "Switch error rate: 0.214000\n",
      "Blockwise Hamming distance: 165\n",
      "Blockwise Hamming rate: 0.147000\n",
      "Largest haplotype segment: 180.471M\n",
      "Most-variants-phased (MVP) segment: 112405\n",
      "\n",
      "=======================chr6=======================\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 110159\n",
      "Percentage of phased SNVs: 96.763\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.763\n",
      "Number of switch errors: 654\n",
      "Switch error rate: 0.594000\n",
      "Blockwise Hamming distance: 450\n",
      "Blockwise Hamming rate: 0.409000\n",
      "Largest haplotype segment: 170.71M\n",
      "Most-variants-phased (MVP) segment: 110159\n",
      "\n",
      "=======================chr7=======================\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 94343\n",
      "Percentage of phased SNVs: 97.053\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.053\n",
      "Number of switch errors: 240\n",
      "Switch error rate: 0.254000\n",
      "Blockwise Hamming distance: 152\n",
      "Blockwise Hamming rate: 0.161000\n",
      "Largest haplotype segment: 159.056M\n",
      "Most-variants-phased (MVP) segment: 94343\n",
      "\n",
      "=======================chr8=======================\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 93248\n",
      "Percentage of phased SNVs: 97.528\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.527\n",
      "Number of switch errors: 88\n",
      "Switch error rate: 0.094000\n",
      "Blockwise Hamming distance: 49\n",
      "Blockwise Hamming rate: 0.053000\n",
      "Largest haplotype segment: 146.137M\n",
      "Most-variants-phased (MVP) segment: 93248\n",
      "\n",
      "=======================chr9=======================\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 67227\n",
      "Percentage of phased SNVs: 96.905\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.905\n",
      "Number of switch errors: 273\n",
      "Switch error rate: 0.406000\n",
      "Blockwise Hamming distance: 294\n",
      "Blockwise Hamming rate: 0.437000\n",
      "Largest haplotype segment: 140.973M\n",
      "Most-variants-phased (MVP) segment: 67227\n",
      "\n",
      "======================chr10=======================\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 86204\n",
      "Percentage of phased SNVs: 97.018\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.018\n",
      "Number of switch errors: 236\n",
      "Switch error rate: 0.274000\n",
      "Blockwise Hamming distance: 147\n",
      "Blockwise Hamming rate: 0.171000\n",
      "Largest haplotype segment: 135.371M\n",
      "Most-variants-phased (MVP) segment: 86204\n",
      "\n",
      "======================chr11=======================\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 84225\n",
      "Percentage of phased SNVs: 97.306\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.306\n",
      "Number of switch errors: 156\n",
      "Switch error rate: 0.185000\n",
      "Blockwise Hamming distance: 101\n",
      "Blockwise Hamming rate: 0.120000\n",
      "Largest haplotype segment: 134.713M\n",
      "Most-variants-phased (MVP) segment: 84225\n",
      "\n",
      "======================chr12=======================\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 75739\n",
      "Percentage of phased SNVs: 97.281\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.281\n",
      "Number of switch errors: 122\n",
      "Switch error rate: 0.161000\n",
      "Blockwise Hamming distance: 80\n",
      "Blockwise Hamming rate: 0.106000\n",
      "Largest haplotype segment: 133.619M\n",
      "Most-variants-phased (MVP) segment: 75739\n",
      "\n",
      "======================chr13=======================\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 64894\n",
      "Percentage of phased SNVs: 97.639\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.639\n",
      "Number of switch errors: 100\n",
      "Switch error rate: 0.154000\n",
      "Blockwise Hamming distance: 57\n",
      "Blockwise Hamming rate: 0.088000\n",
      "Largest haplotype segment: 96.045M\n",
      "Most-variants-phased (MVP) segment: 64894\n",
      "\n",
      "======================chr14=======================\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 50595\n",
      "Percentage of phased SNVs: 97.44\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.44\n",
      "Number of switch errors: 112\n",
      "Switch error rate: 0.221000\n",
      "Blockwise Hamming distance: 65\n",
      "Blockwise Hamming rate: 0.128000\n",
      "Largest haplotype segment: 88.262M\n",
      "Most-variants-phased (MVP) segment: 50595\n",
      "\n",
      "======================chr15=======================\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 45248\n",
      "Percentage of phased SNVs: 94.333\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.333\n",
      "Number of switch errors: 860\n",
      "Switch error rate: 1.901000\n",
      "Blockwise Hamming distance: 685\n",
      "Blockwise Hamming rate: 1.514000\n",
      "Largest haplotype segment: 82.463M\n",
      "Most-variants-phased (MVP) segment: 45248\n",
      "\n",
      "======================chr16=======================\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 43590\n",
      "Percentage of phased SNVs: 95.191\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 95.191\n",
      "Number of switch errors: 526\n",
      "Switch error rate: 1.207000\n",
      "Blockwise Hamming distance: 430\n",
      "Blockwise Hamming rate: 0.986000\n",
      "Largest haplotype segment: 90.052M\n",
      "Most-variants-phased (MVP) segment: 43590\n",
      "\n",
      "======================chr17=======================\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 34920\n",
      "Percentage of phased SNVs: 94.793\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.793\n",
      "Number of switch errors: 212\n",
      "Switch error rate: 0.607000\n",
      "Blockwise Hamming distance: 173\n",
      "Blockwise Hamming rate: 0.495000\n",
      "Largest haplotype segment: 80.874M\n",
      "Most-variants-phased (MVP) segment: 34920\n",
      "\n",
      "======================chr18=======================\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 44455\n",
      "Percentage of phased SNVs: 97.189\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 97.188\n",
      "Number of switch errors: 84\n",
      "Switch error rate: 0.189000\n",
      "Blockwise Hamming distance: 44\n",
      "Blockwise Hamming rate: 0.099000\n",
      "Largest haplotype segment: 77.932M\n",
      "Most-variants-phased (MVP) segment: 44455\n",
      "\n",
      "======================chr19=======================\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 23075\n",
      "Percentage of phased SNVs: 94.772\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.771\n",
      "Number of switch errors: 128\n",
      "Switch error rate: 0.555000\n",
      "Blockwise Hamming distance: 110\n",
      "Blockwise Hamming rate: 0.477000\n",
      "Largest haplotype segment: 58.827M\n",
      "Most-variants-phased (MVP) segment: 23075\n",
      "\n",
      "======================chr20=======================\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 31546\n",
      "Percentage of phased SNVs: 96.107\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.106\n",
      "Number of switch errors: 114\n",
      "Switch error rate: 0.361000\n",
      "Blockwise Hamming distance: 189\n",
      "Blockwise Hamming rate: 0.599000\n",
      "Largest haplotype segment: 62.903M\n",
      "Most-variants-phased (MVP) segment: 31546\n",
      "\n",
      "======================chr21=======================\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 21543\n",
      "Percentage of phased SNVs: 96.727\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 96.727\n",
      "Number of switch errors: 295\n",
      "Switch error rate: 1.369000\n",
      "Blockwise Hamming distance: 193\n",
      "Blockwise Hamming rate: 0.896000\n",
      "Largest haplotype segment: 37.396M\n",
      "Most-variants-phased (MVP) segment: 21543\n",
      "\n",
      "======================chr22=======================\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 14523\n",
      "Percentage of phased SNVs: 94.674\n",
      "Number of phased blocks: 1\n",
      "Completeness (NC): 94.674\n",
      "Number of switch errors: 56\n",
      "Switch error rate: 0.386000\n",
      "Blockwise Hamming distance: 34\n",
      "Blockwise Hamming rate: 0.234000\n",
      "Largest haplotype segment: 35.145M\n",
      "Most-variants-phased (MVP) segment: 14523\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "nSAM = '10'\n",
    "\n",
    "fls = glob.glob('whatshap/*.'+nSAM+'.phased.whatshap.recode.vcf')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[1].replace('chr', '')))\n",
    "\n",
    "numberOfSNPs = OrderedDict()\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[1]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = OrderedDict()\n",
    "    hapPos = OrderedDict()\n",
    "    numberOfSNPs[chrom] = -1\n",
    "\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            numberOfSNPs[chrom] += 1\n",
    "        \n",
    "            if lst[-1] == 'NA' or lst[-1] == 'block':\n",
    "                continue\n",
    "        \n",
    "            if lst[-1] not in hapBlock.keys():\n",
    "                hapBlock[lst[-1]] = []\n",
    "                hapPos[lst[-1]] = []\n",
    "            hapBlock[lst[-1]].append(lst[2])\n",
    "            hapPos[lst[-1]].append(int(lst[1]))\n",
    "    \n",
    "    \n",
    "    numberOfSegments = len(hapBlock)\n",
    "    \n",
    "    phasedSNPs = 0\n",
    "    \n",
    "    mismatch = switchError = hammingDistance = 0\n",
    "    \n",
    "    #hapChr = []\n",
    "    for ke, val in hapBlock.items():\n",
    "        phasedSNPs += len(val)\n",
    "        \n",
    "        if len(Counter(val))==2:\n",
    "            hammingDistance += Counter(val).most_common(2)[1][1]\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        #hapChr.append(len(val))\n",
    "        \n",
    "        v1 = val[1:]\n",
    "        v2 = val[:-1]\n",
    "        \n",
    "        vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "        switchError += sum(vd)\n",
    "\n",
    "        if len(vd)==1:\n",
    "            continue\n",
    "                \n",
    "        indexV = [i+1 for i in range(len(vd)) if vd[i]==True]\n",
    "        indexV1 = indexV[1:]\n",
    "        indexV2 = indexV[:-1]\n",
    "        \n",
    "        indexVd = [a - b for a, b in zip(indexV1,indexV2)]\n",
    "        \n",
    "        mismatch += indexVd.count(1)\n",
    "        \n",
    "        #indexV1.extend([len(val)])\n",
    "        #indexV2.insert(0,0)\n",
    "        #hapChr.extend([a - b for a, b in zip(indexV1, indexV2)])\n",
    "        \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs[chrom]*100,3)\n",
    "        \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "    completeness = round((phasedSNPs-numberOfSegments)/numberOfSNPs[chrom]*100,3)\n",
    "    \n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    #v1 = hapChr[1:]\n",
    "    #v2 = hapChr[:-1]\n",
    "    #error = sum([a != b for a, b in zip(v1,v2)])\n",
    "    #ser = error/(cmplt)*100\n",
    "    \n",
    "    \n",
    "    LHSLength = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = val[-1] - val[0] + 1\n",
    "            \n",
    "        if l > LHSLength:\n",
    "            LHSLength = l\n",
    "            LHSVariant = len(val)\n",
    "            \n",
    "    LHSLength = round(LHSLength/1000000,3)\n",
    "        \n",
    "    for ke, val in hapPos.items():\n",
    "        l = round((val[-1] - val[0] + 1)/1000000,3)\n",
    "        \n",
    "        if l == LHSLength:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    MVP = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = len(val)\n",
    "            \n",
    "        if l > MVP:\n",
    "            MVP = l\n",
    "    \n",
    "    for ke, val in hapPos.items():\n",
    "        if len(val) == MVP:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    #QAN50 = 0\n",
    "    #for size in sorted(hapChr, reverse=True):\n",
    "    #    QAN50 += size\n",
    "        #print ('largest %s, %s' %(size, QAN50))\n",
    "    \n",
    "    \n",
    "        #if QAN50 > phased/2:\n",
    "            #QAN50 = size\n",
    "            #print ('QAN %s' %size)\n",
    "            #break\n",
    "    \n",
    "        #ratio = Counter(val).most_common(1)[0][1]/len(val)\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        \n",
    "    metrics[chrom] = [numberOfSNPs[chrom], phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs[chrom])\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()\n",
    "    \n",
    "    \n",
    "fh = open('whatshap/metrics.whatshap.'+nSAM+'.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('whatshap/*.'+nSAM+'.phased.HAPCUT2.recode.vcf')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[1].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[1]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = OrderedDict()\n",
    "    hapPos = OrderedDict()\n",
    "    #numberOfSNPs = -1\n",
    "\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            #numberOfSNPs += 1\n",
    "        \n",
    "            if lst[-1] == 'NA' or lst[-1] == 'block':\n",
    "                continue\n",
    "        \n",
    "            if lst[-1] not in hapBlock.keys():\n",
    "                hapBlock[lst[-1]] = []\n",
    "                hapPos[lst[-1]] = []\n",
    "            hapBlock[lst[-1]].append(lst[2])\n",
    "            hapPos[lst[-1]].append(int(lst[1]))\n",
    "    \n",
    "    \n",
    "    numberOfSegments = len(hapBlock)\n",
    "    \n",
    "    phasedSNPs = 0\n",
    "    \n",
    "    mismatch = switchError = hammingDistance = 0\n",
    "    \n",
    "    #hapChr = []\n",
    "    for ke, val in hapBlock.items():\n",
    "        phasedSNPs += len(val)\n",
    "        \n",
    "        if len(Counter(val))==2:\n",
    "            hammingDistance += Counter(val).most_common(2)[1][1]\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        #hapChr.append(len(val))\n",
    "        \n",
    "        v1 = val[1:]\n",
    "        v2 = val[:-1]\n",
    "        \n",
    "        vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "        switchError += sum(vd)\n",
    "\n",
    "        if len(vd)==1:\n",
    "            continue\n",
    "                \n",
    "        indexV = [i+1 for i in range(len(vd)) if vd[i]==True]\n",
    "        indexV1 = indexV[1:]\n",
    "        indexV2 = indexV[:-1]\n",
    "        \n",
    "        indexVd = [a - b for a, b in zip(indexV1,indexV2)]\n",
    "        \n",
    "        mismatch += indexVd.count(1)\n",
    "        \n",
    "        #indexV1.extend([len(val)])\n",
    "        #indexV2.insert(0,0)\n",
    "        #hapChr.extend([a - b for a, b in zip(indexV1, indexV2)])\n",
    "        \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs[chrom]*100,3)\n",
    "        \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "    completeness = round((phasedSNPs-numberOfSegments)/numberOfSNPs[chrom]*100,3)\n",
    "    \n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    #v1 = hapChr[1:]\n",
    "    #v2 = hapChr[:-1]\n",
    "    #error = sum([a != b for a, b in zip(v1,v2)])\n",
    "    #ser = error/(cmplt)*100\n",
    "    \n",
    "    \n",
    "    LHSLength = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = val[-1] - val[0] + 1\n",
    "            \n",
    "        if l > LHSLength:\n",
    "            LHSLength = l\n",
    "            LHSVariant = len(val)\n",
    "            \n",
    "    LHSLength = round(LHSLength/1000000,3)\n",
    "        \n",
    "    for ke, val in hapPos.items():\n",
    "        l = round((val[-1] - val[0] + 1)/1000000,3)\n",
    "        \n",
    "        if l == LHSLength:\n",
    "            print (ke)\n",
    "        \n",
    "        \n",
    "    MVP = 0\n",
    "    for ke, val in hapPos.items():\n",
    "        l = len(val)\n",
    "            \n",
    "        if l > MVP:\n",
    "            MVP = l\n",
    "            \n",
    "    for ke, val in hapPos.items():\n",
    "        if len(val) == MVP:\n",
    "            print (ke)\n",
    "    \n",
    "    \n",
    "    #QAN50 = 0\n",
    "    #for size in sorted(hapChr, reverse=True):\n",
    "    #    QAN50 += size\n",
    "        #print ('largest %s, %s' %(size, QAN50))\n",
    "    \n",
    "    \n",
    "        #if QAN50 > phased/2:\n",
    "            #QAN50 = size\n",
    "            #print ('QAN %s' %size)\n",
    "            #break\n",
    "    \n",
    "        #ratio = Counter(val).most_common(1)[0][1]/len(val)\n",
    "        #if (ratio < 1):\n",
    "        #    i += Counter(val).most_common(2)[1][1]\n",
    "        #print (i)\n",
    "        \n",
    "        \n",
    "    metrics[chrom] = [numberOfSNPs[chrom], phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs[chrom])\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()\n",
    "    \n",
    "    \n",
    "fh = open('whatshap/metrics.HAPCUT2.'+nSAM+'.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('whatshap/NEW5.Hapi2.sperm'+nSAM+'.chr*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[3].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[3]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = []\n",
    "    hapPos = []\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            \n",
    "            if line.startswith('hap1'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            hapBlock.append(lst[1])\n",
    "            hapPos.append(lst[0].split(':')[1])\n",
    "            #hapPos.append(lst[0])\n",
    "    \n",
    "    \n",
    "    numberOfSegments = 1\n",
    "    #numberOfSNPs = len(hapBlock)\n",
    "    \n",
    "    hapCount = Counter(hapBlock)\n",
    "    \n",
    "    phasedSNPs = hapCount['0']+hapCount['1']\n",
    "    completeness = round((phasedSNPs-1)/(numberOfSNPs[chrom]-1)*100,3)\n",
    "    \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs[chrom]*100,3)\n",
    "    \n",
    "    hammingDistance = min(hapCount['0'], hapCount['1'])\n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    LHSLength = round((int(hapPos[-1])-int(hapPos[0]))/1000000,3)\n",
    "    MVP = phasedSNPs\n",
    "    \n",
    "    hapClean = [hap for hap in hapBlock if hap != '7']\n",
    "    \n",
    "    v1 = hapClean[1:]\n",
    "    v2 = hapClean[:-1]\n",
    "        \n",
    "    vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "    switchError = sum(vd)\n",
    "    \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "     \n",
    "    metrics[chrom] = [numberOfSNPs[chrom], phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs[chrom])\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()\n",
    "\n",
    "    \n",
    "fh = open('whatshap/metrics.Hapi.'+nSAM+'.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('whatshap/Cell2.sperm'+nSAM+'.chr*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[2].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[2]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = []\n",
    "    hapPos = []\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            \n",
    "            if line.startswith('hap1'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            hapBlock.append(lst[1])\n",
    "            #hapPos.append(lst[0].split(':')[1])\n",
    "            hapPos.append(lst[0])\n",
    "    \n",
    "    \n",
    "    numberOfSegments = 1\n",
    "    #numberOfSNPs = len(hapBlock)\n",
    "    \n",
    "    hapCount = Counter(hapBlock)\n",
    "    \n",
    "    phasedSNPs = hapCount['0']+hapCount['1']\n",
    "    completeness = round((phasedSNPs-1)/(numberOfSNPs[chrom]-1)*100,3)\n",
    "    \n",
    "    phasedPercent = round(phasedSNPs/numberOfSNPs[chrom]*100,3)\n",
    "    \n",
    "    hammingDistance = min(hapCount['0'], hapCount['1'])\n",
    "    hammingRate = round(hammingDistance/phasedSNPs*100,3)\n",
    "    \n",
    "    LHSLength = round((int(hapPos[-1])-int(hapPos[0]))/1000000,3)\n",
    "    MVP = phasedSNPs\n",
    "    \n",
    "    hapClean = [hap for hap in hapBlock if hap != '7']\n",
    "    \n",
    "    v1 = hapClean[1:]\n",
    "    v2 = hapClean[:-1]\n",
    "        \n",
    "    vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "    switchError = sum(vd)\n",
    "    \n",
    "    SER = round(switchError/(phasedSNPs-numberOfSegments)*100,3)\n",
    "     \n",
    "    metrics[chrom] = [numberOfSNPs[chrom], phasedSNPs, phasedPercent, numberOfSegments,\n",
    "                      completeness, switchError, SER, hammingDistance, hammingRate,\n",
    "                      LHSLength, MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %numberOfSNPs[chrom])\n",
    "    print ('Number of phased SNVs: %s' %phasedSNPs)\n",
    "    print ('Percentage of phased SNVs: %s' %phasedPercent)\n",
    "    print ('Number of phased blocks: %s' %numberOfSegments)\n",
    "    print ('Completeness (NC): %s' %completeness)\n",
    "    \n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %switchError)\n",
    "    print ('Switch error rate: %f' %SER)\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hammingDistance)\n",
    "    print ('Blockwise Hamming rate: %f' %hammingRate)\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %LHSLength)\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()\n",
    "    \n",
    "\n",
    "fh = open('whatshap/metrics.PHMM.'+nSAM+'.txt', 'wt')\n",
    "\n",
    "fh.write('\\t'.join(['numberOfSNPs','phasedSNPs','phasedPercent',\n",
    "                   'numberOfSegments','completeness','switchError',\n",
    "                   'SER','hammingDistance','hammingRate','LHSLength','MVP'])+'\\n')\n",
    "\n",
    "for ke, val in metrics.items():\n",
    "    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr1.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr2.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr3.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr4.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr5.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr6.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr7.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr8.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr9.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr10.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr11.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr12.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr13.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr14.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr15.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr16.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr17.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr18.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr19.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr20.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr21.haplotype.all.txt',\n",
       " 'Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr22.haplotype.all.txt']"
      ]
     },
     "execution_count": 400,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fls = glob.glob('Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[3].replace('chr', '')))\n",
    "fls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {},
   "outputs": [],
   "source": [
    "hapBlock = []\n",
    "hapPos = []\n",
    "with open('Comparison/GenomeResearch/Consistency/NEW5.Hapi2.sperm11.chr1.haplotype.all.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('hap1'):\n",
    "            continue\n",
    "        \n",
    "        lst = line.split('\\t')\n",
    "            \n",
    "        hapBlock.append(lst[1])\n",
    "        hapPos.append(lst[0].split(':')[1])\n",
    "    \n",
    "tot = len(hapBlock)\n",
    "hapCount = Counter(hapBlock)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129720"
      ]
     },
     "execution_count": 414,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "479"
      ]
     },
     "execution_count": 413,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapCount['7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "315"
      ]
     },
     "execution_count": 424,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(hapCount['0'], hapCount['1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "128926"
      ]
     },
     "execution_count": 425,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapCount['0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fls = glob.glob('Comparison/GenomeResearch/Consistency/Cell2.sperm11.*.haplotype.all')\n",
    "fls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======================chr1=======================\n",
      "Total number of SNPs: 129720\n",
      "Number of phased SNVs: 120708\n",
      "Percentage of phased SNVs: 0.931\n",
      "Completeness (NC): 0.931\n",
      "Number of switch errors: 4623\n",
      "Switch error rate: 3.830000\n",
      "Blockwise Hamming distance: 6790\n",
      "Blockwise Hamming rate: 5.625000\n",
      "Largest haplotype segment: 248.393M\n",
      "Most-variants-phased (MVP) segment: 120708\n",
      "\n",
      "=======================chr2=======================\n",
      "Total number of SNPs: 141669\n",
      "Number of phased SNVs: 138930\n",
      "Percentage of phased SNVs: 0.981\n",
      "Completeness (NC): 0.981\n",
      "Number of switch errors: 1298\n",
      "Switch error rate: 0.934000\n",
      "Blockwise Hamming distance: 1260\n",
      "Blockwise Hamming rate: 0.907000\n",
      "Largest haplotype segment: 242.959M\n",
      "Most-variants-phased (MVP) segment: 138930\n",
      "\n",
      "=======================chr3=======================\n",
      "Total number of SNPs: 127247\n",
      "Number of phased SNVs: 127067\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 294\n",
      "Switch error rate: 0.231000\n",
      "Blockwise Hamming distance: 202\n",
      "Blockwise Hamming rate: 0.159000\n",
      "Largest haplotype segment: 197.794M\n",
      "Most-variants-phased (MVP) segment: 127067\n",
      "\n",
      "=======================chr4=======================\n",
      "Total number of SNPs: 130152\n",
      "Number of phased SNVs: 129980\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 438\n",
      "Switch error rate: 0.337000\n",
      "Blockwise Hamming distance: 546\n",
      "Blockwise Hamming rate: 0.420000\n",
      "Largest haplotype segment: 191.009M\n",
      "Most-variants-phased (MVP) segment: 129980\n",
      "\n",
      "=======================chr5=======================\n",
      "Total number of SNPs: 114988\n",
      "Number of phased SNVs: 114770\n",
      "Percentage of phased SNVs: 0.998\n",
      "Completeness (NC): 0.998\n",
      "Number of switch errors: 245\n",
      "Switch error rate: 0.213000\n",
      "Blockwise Hamming distance: 165\n",
      "Blockwise Hamming rate: 0.144000\n",
      "Largest haplotype segment: 180.652M\n",
      "Most-variants-phased (MVP) segment: 114770\n",
      "\n",
      "=======================chr6=======================\n",
      "Total number of SNPs: 113844\n",
      "Number of phased SNVs: 113218\n",
      "Percentage of phased SNVs: 0.995\n",
      "Completeness (NC): 0.995\n",
      "Number of switch errors: 642\n",
      "Switch error rate: 0.567000\n",
      "Blockwise Hamming distance: 445\n",
      "Blockwise Hamming rate: 0.393000\n",
      "Largest haplotype segment: 170.71M\n",
      "Most-variants-phased (MVP) segment: 113218\n",
      "\n",
      "=======================chr7=======================\n",
      "Total number of SNPs: 97208\n",
      "Number of phased SNVs: 97029\n",
      "Percentage of phased SNVs: 0.998\n",
      "Completeness (NC): 0.998\n",
      "Number of switch errors: 250\n",
      "Switch error rate: 0.258000\n",
      "Blockwise Hamming distance: 169\n",
      "Blockwise Hamming rate: 0.174000\n",
      "Largest haplotype segment: 159.066M\n",
      "Most-variants-phased (MVP) segment: 97029\n",
      "\n",
      "=======================chr8=======================\n",
      "Total number of SNPs: 95612\n",
      "Number of phased SNVs: 95561\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 82\n",
      "Switch error rate: 0.086000\n",
      "Blockwise Hamming distance: 46\n",
      "Blockwise Hamming rate: 0.048000\n",
      "Largest haplotype segment: 146.137M\n",
      "Most-variants-phased (MVP) segment: 95561\n",
      "\n",
      "=======================chr9=======================\n",
      "Total number of SNPs: 69374\n",
      "Number of phased SNVs: 69159\n",
      "Percentage of phased SNVs: 0.997\n",
      "Completeness (NC): 0.997\n",
      "Number of switch errors: 275\n",
      "Switch error rate: 0.398000\n",
      "Blockwise Hamming distance: 294\n",
      "Blockwise Hamming rate: 0.425000\n",
      "Largest haplotype segment: 140.973M\n",
      "Most-variants-phased (MVP) segment: 69159\n",
      "\n",
      "======================chr10=======================\n",
      "Total number of SNPs: 88854\n",
      "Number of phased SNVs: 88668\n",
      "Percentage of phased SNVs: 0.998\n",
      "Completeness (NC): 0.998\n",
      "Number of switch errors: 236\n",
      "Switch error rate: 0.266000\n",
      "Blockwise Hamming distance: 144\n",
      "Blockwise Hamming rate: 0.162000\n",
      "Largest haplotype segment: 135.371M\n",
      "Most-variants-phased (MVP) segment: 88668\n",
      "\n",
      "======================chr11=======================\n",
      "Total number of SNPs: 86557\n",
      "Number of phased SNVs: 86463\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 158\n",
      "Switch error rate: 0.183000\n",
      "Blockwise Hamming distance: 101\n",
      "Blockwise Hamming rate: 0.117000\n",
      "Largest haplotype segment: 134.713M\n",
      "Most-variants-phased (MVP) segment: 86463\n",
      "\n",
      "======================chr12=======================\n",
      "Total number of SNPs: 77856\n",
      "Number of phased SNVs: 77794\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 124\n",
      "Switch error rate: 0.159000\n",
      "Blockwise Hamming distance: 75\n",
      "Blockwise Hamming rate: 0.096000\n",
      "Largest haplotype segment: 133.619M\n",
      "Most-variants-phased (MVP) segment: 77794\n",
      "\n",
      "======================chr13=======================\n",
      "Total number of SNPs: 66463\n",
      "Number of phased SNVs: 66394\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 96\n",
      "Switch error rate: 0.145000\n",
      "Blockwise Hamming distance: 54\n",
      "Blockwise Hamming rate: 0.081000\n",
      "Largest haplotype segment: 96.045M\n",
      "Most-variants-phased (MVP) segment: 66394\n",
      "\n",
      "======================chr14=======================\n",
      "Total number of SNPs: 51924\n",
      "Number of phased SNVs: 51860\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 102\n",
      "Switch error rate: 0.197000\n",
      "Blockwise Hamming distance: 57\n",
      "Blockwise Hamming rate: 0.110000\n",
      "Largest haplotype segment: 88.262M\n",
      "Most-variants-phased (MVP) segment: 51860\n",
      "\n",
      "======================chr15=======================\n",
      "Total number of SNPs: 47966\n",
      "Number of phased SNVs: 46688\n",
      "Percentage of phased SNVs: 0.973\n",
      "Completeness (NC): 0.973\n",
      "Number of switch errors: 776\n",
      "Switch error rate: 1.662000\n",
      "Blockwise Hamming distance: 689\n",
      "Blockwise Hamming rate: 1.476000\n",
      "Largest haplotype segment: 82.463M\n",
      "Most-variants-phased (MVP) segment: 46688\n",
      "\n",
      "======================chr16=======================\n",
      "Total number of SNPs: 45792\n",
      "Number of phased SNVs: 45282\n",
      "Percentage of phased SNVs: 0.989\n",
      "Completeness (NC): 0.989\n",
      "Number of switch errors: 520\n",
      "Switch error rate: 1.148000\n",
      "Blockwise Hamming distance: 443\n",
      "Blockwise Hamming rate: 0.978000\n",
      "Largest haplotype segment: 90.057M\n",
      "Most-variants-phased (MVP) segment: 45282\n",
      "\n",
      "======================chr17=======================\n",
      "Total number of SNPs: 36838\n",
      "Number of phased SNVs: 36656\n",
      "Percentage of phased SNVs: 0.995\n",
      "Completeness (NC): 0.995\n",
      "Number of switch errors: 192\n",
      "Switch error rate: 0.524000\n",
      "Blockwise Hamming distance: 151\n",
      "Blockwise Hamming rate: 0.412000\n",
      "Largest haplotype segment: 80.874M\n",
      "Most-variants-phased (MVP) segment: 36656\n",
      "\n",
      "======================chr18=======================\n",
      "Total number of SNPs: 45741\n",
      "Number of phased SNVs: 45695\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 80\n",
      "Switch error rate: 0.175000\n",
      "Blockwise Hamming distance: 40\n",
      "Blockwise Hamming rate: 0.088000\n",
      "Largest haplotype segment: 77.932M\n",
      "Most-variants-phased (MVP) segment: 45695\n",
      "\n",
      "======================chr19=======================\n",
      "Total number of SNPs: 24348\n",
      "Number of phased SNVs: 24259\n",
      "Percentage of phased SNVs: 0.996\n",
      "Completeness (NC): 0.996\n",
      "Number of switch errors: 118\n",
      "Switch error rate: 0.486000\n",
      "Blockwise Hamming distance: 102\n",
      "Blockwise Hamming rate: 0.420000\n",
      "Largest haplotype segment: 58.827M\n",
      "Most-variants-phased (MVP) segment: 24259\n",
      "\n",
      "======================chr20=======================\n",
      "Total number of SNPs: 32824\n",
      "Number of phased SNVs: 32684\n",
      "Percentage of phased SNVs: 0.996\n",
      "Completeness (NC): 0.996\n",
      "Number of switch errors: 110\n",
      "Switch error rate: 0.337000\n",
      "Blockwise Hamming distance: 179\n",
      "Blockwise Hamming rate: 0.548000\n",
      "Largest haplotype segment: 62.903M\n",
      "Most-variants-phased (MVP) segment: 32684\n",
      "\n",
      "======================chr21=======================\n",
      "Total number of SNPs: 22272\n",
      "Number of phased SNVs: 22087\n",
      "Percentage of phased SNVs: 0.992\n",
      "Completeness (NC): 0.992\n",
      "Number of switch errors: 277\n",
      "Switch error rate: 1.254000\n",
      "Blockwise Hamming distance: 172\n",
      "Blockwise Hamming rate: 0.779000\n",
      "Largest haplotype segment: 37.396M\n",
      "Most-variants-phased (MVP) segment: 22087\n",
      "\n",
      "======================chr22=======================\n",
      "Total number of SNPs: 15340\n",
      "Number of phased SNVs: 15323\n",
      "Percentage of phased SNVs: 0.999\n",
      "Completeness (NC): 0.999\n",
      "Number of switch errors: 48\n",
      "Switch error rate: 0.313000\n",
      "Blockwise Hamming distance: 31\n",
      "Blockwise Hamming rate: 0.202000\n",
      "Largest haplotype segment: 35.145M\n",
      "Most-variants-phased (MVP) segment: 15323\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Number of phased blocks\n",
    "### Percentage of phased single nucleotide variants (SNVs)\n",
    "\n",
    "fls = glob.glob('Comparison/GenomeResearch/Consistency/Cell2.sperm11.*.haplotype.all.txt')\n",
    "fls = sorted(fls, key = lambda x:int(x.split('.')[2].replace('chr', '')))\n",
    "\n",
    "\n",
    "metrics = OrderedDict()\n",
    "for fl in fls:\n",
    "    chrom = fl.split('.')[2]\n",
    "    \n",
    "    print (format(chrom, '=^50'))\n",
    "    \n",
    "    hapBlock = []\n",
    "    hapPos = []\n",
    "    with open(fl) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            \n",
    "            if line.startswith('hap1'):\n",
    "                continue\n",
    "            \n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            hapBlock.append(lst[1])\n",
    "            #hapPos.append(lst[0].split(':')[1])\n",
    "            hapPos.append(lst[0])\n",
    "    \n",
    "    \n",
    "    tot = len(hapBlock)\n",
    "    hapCount = Counter(hapBlock)\n",
    "    phased = tot-hapCount['7']\n",
    "    \n",
    "    completeness = (phased-1)/(tot-1)\n",
    "    \n",
    "    hamming = min(hapCount['0'], hapCount['1'])\n",
    "    \n",
    "    LPSLength = (int(hapPos[-1])-int(hapPos[0]))/1000000\n",
    "    MVP = phased\n",
    "    \n",
    "    hapClean = [hap for hap in hapBlock if hap != '7']\n",
    "    \n",
    "    v1 = hapClean[1:]\n",
    "    v2 = hapClean[:-1]\n",
    "        \n",
    "    vd = [a != b for a, b in zip(v1,v2)]\n",
    "        \n",
    "    error = sum(vd)\n",
    "    \n",
    "    ser = error/(phased-1)*100\n",
    "     \n",
    "    metrics[chrom] = [tot, phased, round(phased/tot*100,3), round(completeness*100,3), \n",
    "                      len(hapBlock), error, round(ser,3), hamming, \n",
    "                      round(hamming/phased*100,3),round(LPSLength,3), MVP]\n",
    "    \n",
    "    ### completeness\n",
    "    print ('Total number of SNPs: %s' %tot)\n",
    "    print ('Number of phased SNVs: %s' %phased)\n",
    "    print ('Percentage of phased SNVs: %s' %round(phased/tot,3))\n",
    "    print ('Completeness (NC): %s' %round(completeness,3))\n",
    "    \n",
    "    ### accuracy\n",
    "    print ('Number of switch errors: %s' %error)\n",
    "    print ('Switch error rate: %f' %round(ser,3))\n",
    "    #print ('Number of mismatch: %s' %mismatch)\n",
    "    \n",
    "    print ('Blockwise Hamming distance: %s' %hamming)\n",
    "    print ('Blockwise Hamming rate: %f' %round(hamming/phased*100,3))\n",
    "    \n",
    "    print ('Largest haplotype segment: %sM' %round(LPSLength,3))\n",
    "    print ('Most-variants-phased (MVP) segment: %s' %MVP)\n",
    "    \n",
    "    #print ('QAN %s' %QAN50)\n",
    "    print ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 361,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(v).most_common(2)[1][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Counter(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78948"
      ]
     },
     "execution_count": 381,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3682,\n",
       " 427,\n",
       " 347,\n",
       " 280,\n",
       " 238,\n",
       " 201,\n",
       " 167,\n",
       " 156,\n",
       " 139,\n",
       " 132,\n",
       " 131,\n",
       " 122,\n",
       " 121,\n",
       " 113,\n",
       " 108,\n",
       " 108,\n",
       " 94,\n",
       " 93,\n",
       " 91,\n",
       " 89,\n",
       " 85,\n",
       " 84,\n",
       " 78,\n",
       " 77,\n",
       " 74,\n",
       " 72,\n",
       " 69,\n",
       " 68,\n",
       " 67,\n",
       " 64,\n",
       " 63,\n",
       " 62,\n",
       " 61,\n",
       " 61,\n",
       " 59,\n",
       " 57,\n",
       " 56,\n",
       " 55,\n",
       " 54,\n",
       " 53,\n",
       " 51,\n",
       " 51,\n",
       " 51,\n",
       " 51,\n",
       " 50,\n",
       " 50,\n",
       " 48,\n",
       " 48,\n",
       " 47,\n",
       " 46,\n",
       " 46,\n",
       " 44,\n",
       " 44,\n",
       " 44,\n",
       " 44,\n",
       " 44,\n",
       " 44,\n",
       " 42,\n",
       " 41,\n",
       " 41,\n",
       " 41,\n",
       " 40,\n",
       " 40,\n",
       " 40,\n",
       " 40,\n",
       " 40,\n",
       " 40,\n",
       " 39,\n",
       " 39,\n",
       " 39,\n",
       " 39,\n",
       " 38,\n",
       " 38,\n",
       " 38,\n",
       " 38,\n",
       " 37,\n",
       " 36,\n",
       " 36,\n",
       " 36,\n",
       " 36,\n",
       " 36,\n",
       " 36,\n",
       " 36,\n",
       " 35,\n",
       " 35,\n",
       " 35,\n",
       " 34,\n",
       " 34,\n",
       " 34,\n",
       " 34,\n",
       " 34,\n",
       " 34,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 33,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 32,\n",
       " 31,\n",
       " 31,\n",
       " 31,\n",
       " 31,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 30,\n",
       " 29,\n",
       " 29,\n",
       " 29,\n",
       " 29,\n",
       " 28,\n",
       " 28,\n",
       " 28,\n",
       " 28,\n",
       " 27,\n",
       " 27,\n",
       " 27,\n",
       " 27,\n",
       " 27,\n",
       " 27,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 26,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 25,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 24,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 23,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 22,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 21,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 20,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 19,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 18,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 17,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 16,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 15,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 14,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 13,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 12,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 11,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 10,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " 9,\n",
       " ...]"
      ]
     },
     "execution_count": 380,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(hapChr, reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "v = [0,0,0,0,0,0,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "v1 = v[1:]\n",
    "v2 = v[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 0]"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1]"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "v = [a != b for a, b in zip(v1,v2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[True, True]"
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [],
   "source": [
    "indexV = [i+1 for i in range(len(v)) if v[i]==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [],
   "source": [
    "indexV1 = indexV[1:]\n",
    "indexV2 = indexV[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1]"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexVd = [a - b for a, b in zip(indexV1,indexV2)]\n",
    "indexVd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {},
   "outputs": [],
   "source": [
    "indexV1.extend([len(v)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 2]"
      ]
     },
     "execution_count": 313,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexV1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {},
   "outputs": [],
   "source": [
    "indexV2.insert(0,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1]"
      ]
     },
     "execution_count": 315,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexV2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([a != b for a, b in zip(v1,v2)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'a' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-149-8da3a14e39b0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;33m[\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m-\u001b[0m\u001b[0mb\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mv1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mv2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'a' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<zip at 0x7f72743ff648>"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "*************************************************************************\n",
    "\n",
    "###### Completeness\n",
    "# % of phased SNVs\n",
    "\n",
    "# sum the number of phase connections across all haplotype segments and divide\n",
    "# by the maximum possible number of phase connections, which is equal to the\n",
    "# number of heterozygous variants on a chromosome minus one.\n",
    "\n",
    "\n",
    "###### Accuracy\n",
    "# switch error rate\n",
    "# blockwise hamming rate\n",
    "\n",
    "\n",
    "###### Contiguity\n",
    "# length of the largest haplotype segment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###### Comparison of phasing strategies for whole human genomes\n",
    "\n",
    "### Percentage of phased single nucleotide variants (SNVs) ***\n",
    "    ### Number of phasing blocks\n",
    "    ### Number of phased SNVs\n",
    "    ### % phased SNPs (among golden set)\n",
    "    \n",
    "### Switch error rate (SER)\n",
    "    ### SNPs used for switch error measurement\n",
    "    ### Number of switch errors\n",
    "    ### SER\n",
    "    \n",
    "### Haplotype block length : QAN50 - Completeness and accuracy\n",
    "\n",
    "\n",
    "### Pairwise SNVs phasing accuracy and phasing yield\n",
    "### Percentage of genes fully phased\n",
    "### SER as a function of haplotype diversity\n",
    "### Shared switch error sites\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###### WhatsHap: fast and accurate read-based phasing\n",
    "\n",
    "### phasing accuracy\n",
    "### phasing completeness \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "###### Dense and accurate whole-chromosome haplotyping of individual genomes\n",
    "\n",
    "### Largest haplotype segment ***\n",
    "\n",
    "### Largest haplotype segment Hamming rate\n",
    "\n",
    "### Hamming error rate is equal to the fraction of wrongly phased heterozygous variants\n",
    "\n",
    "\n",
    "# completeness (i.e., % of phased SNVs), \n",
    "# contiguity (length of the largest haplotype segment) ***\n",
    "# accuracy (agreement with the “reference” standard) of each assembled haplotype.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('16096825', ['1', '1']),\n",
       "             ('16106307', ['0', '1', '1']),\n",
       "             ('16191691', ['1', '1']),\n",
       "             ('16228793', ['1', '1']),\n",
       "             ('16874056', ['0', '0', '0', '1']),\n",
       "             ('16876292', ['1', '1', '1']),\n",
       "             ('16888538', ['1', '1', '1']),\n",
       "             ('16890136', ['1', '1']),\n",
       "             ('16894117', ['1', '0']),\n",
       "             ('16906273', ['1', '1']),\n",
       "             ('17066493', ['1', '1', '1']),\n",
       "             ('17107266', ['0', '0']),\n",
       "             ('17108394', ['1', '1']),\n",
       "             ('17118076', ['1', '1', '1', '1', '1']),\n",
       "             ('17122332', ['1', '1']),\n",
       "             ('17130268', ['1', '1', '1']),\n",
       "             ('17141339', ['1', '1']),\n",
       "             ('17150676', ['1', '1']),\n",
       "             ('17152611', ['1', '1']),\n",
       "             ('17160807', ['1', '1', '1']),\n",
       "             ('17169347', ['1', '1']),\n",
       "             ('17169979', ['1', '1']),\n",
       "             ('17175781', ['1', '1']),\n",
       "             ('17177906', ['1', '1']),\n",
       "             ('17234497', ['1', '1']),\n",
       "             ('17238335', ['1', '1']),\n",
       "             ('17295703', ['1', '1']),\n",
       "             ('17296426',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('17301217', ['1', '1', '1', '1']),\n",
       "             ('17307742', ['1', '1', '1']),\n",
       "             ('17310488', ['0', '0']),\n",
       "             ('17312596', ['1', '1']),\n",
       "             ('17369658', ['1', '1']),\n",
       "             ('17388093', ['1', '1']),\n",
       "             ('17388723', ['1', '1', '1']),\n",
       "             ('17398508', ['0', '0']),\n",
       "             ('17404605', ['1', '1']),\n",
       "             ('17406581', ['1', '1']),\n",
       "             ('17407349', ['1', '1', '1']),\n",
       "             ('17409813', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('17411899', ['1', '1', '1', '1', '1']),\n",
       "             ('17492452', ['1', '1']),\n",
       "             ('17493293', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('17502191', ['0', '1', '0']),\n",
       "             ('17503328', ['1', '0']),\n",
       "             ('17504553', ['0', '1']),\n",
       "             ('17506199', ['1', '1', '1']),\n",
       "             ('17511308', ['0', '0', '0']),\n",
       "             ('17513703', ['1', '0', '0']),\n",
       "             ('17514686', ['0', '1']),\n",
       "             ('17525584', ['1', '1']),\n",
       "             ('17526489', ['1', '1', '1', '1', '1']),\n",
       "             ('17529583', ['0', '0']),\n",
       "             ('17530370', ['1', '1']),\n",
       "             ('17532449', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('17534726', ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('17536145', ['1', '1', '1']),\n",
       "             ('17537530',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('17539427', ['1', '1', '1']),\n",
       "             ('17542810', ['0', '0']),\n",
       "             ('17548971', ['1', '1']),\n",
       "             ('17549526', ['0', '0']),\n",
       "             ('17556720', ['1', '1', '1', '1']),\n",
       "             ('17564907', ['1', '1', '1']),\n",
       "             ('17565888', ['1', '1']),\n",
       "             ('17567699', ['1', '1', '1']),\n",
       "             ('17597297', ['1', '1', '1']),\n",
       "             ('17603445', ['1', '1', '1']),\n",
       "             ('17606774', ['1', '1']),\n",
       "             ('17612518', ['1', '1', '1']),\n",
       "             ('17614011', ['1', '1', '1', '1', '1']),\n",
       "             ('17615851', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('17619292', ['1', '1', '1', '1']),\n",
       "             ('17622267', ['0', '0']),\n",
       "             ('17623837', ['1', '1']),\n",
       "             ('17628648', ['0', '0']),\n",
       "             ('17630389', ['1', '1']),\n",
       "             ('17631691', ['0', '0', '0']),\n",
       "             ('17637975', ['0', '0', '0']),\n",
       "             ('17639837', ['1', '1']),\n",
       "             ('17642094', ['1', '1']),\n",
       "             ('17649549', ['0', '0']),\n",
       "             ('17650705', ['0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('17652955', ['1', '1']),\n",
       "             ('17653927', ['1', '1']),\n",
       "             ('17654790', ['0', '0', '0', '0', '0']),\n",
       "             ('17655995', ['0', '0', '0', '0']),\n",
       "             ('17657292', ['1', '1']),\n",
       "             ('17658341', ['1', '1', '1', '1', '1']),\n",
       "             ('17659579', ['0', '0', '0', '0', '0']),\n",
       "             ('17661508', ['1', '1']),\n",
       "             ('17661764', ['1', '1']),\n",
       "             ('17668818', ['1', '1', '1']),\n",
       "             ('17671052', ['1', '1']),\n",
       "             ('17677892', ['0', '0', '0', '0']),\n",
       "             ('17696027', ['0', '0']),\n",
       "             ('17697861', ['1', '1']),\n",
       "             ('17702580', ['1', '1']),\n",
       "             ('17706647', ['1', '1']),\n",
       "             ('17707727', ['1', '1']),\n",
       "             ('17734704', ['1', '1']),\n",
       "             ('17735943', ['1', '1', '1', '1']),\n",
       "             ('17743228', ['1', '1', '1', '1']),\n",
       "             ('17747278', ['1', '1']),\n",
       "             ('17790611', ['1', '1']),\n",
       "             ('17800472', ['1', '1']),\n",
       "             ('17803593', ['1', '1', '1', '1', '1']),\n",
       "             ('17805272', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('17825248', ['1', '1']),\n",
       "             ('17826017', ['1', '1']),\n",
       "             ('17827684', ['1', '1']),\n",
       "             ('17829248', ['1', '1']),\n",
       "             ('17831813', ['0', '0', '0', '0']),\n",
       "             ('17833100', ['1', '1']),\n",
       "             ('17834460', ['1', '1']),\n",
       "             ('17840229', ['1', '1']),\n",
       "             ('17850483', ['1', '1', '1']),\n",
       "             ('17851410', ['1', '1']),\n",
       "             ('17853714', ['1', '1']),\n",
       "             ('17856770', ['1', '1']),\n",
       "             ('17858175', ['1', '1']),\n",
       "             ('17860048', ['1', '1']),\n",
       "             ('17861981', ['1', '1']),\n",
       "             ('17932686', ['1', '1']),\n",
       "             ('17939816', ['0', '0']),\n",
       "             ('17942872', ['1', '1']),\n",
       "             ('17944167', ['1', '1']),\n",
       "             ('17948335', ['1', '1', '1', '1']),\n",
       "             ('17965531', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('17968155', ['1', '1', '1']),\n",
       "             ('17973379', ['0', '0']),\n",
       "             ('17977024', ['1', '1']),\n",
       "             ('17978090', ['1', '1', '1']),\n",
       "             ('17984815', ['0', '0', '0', '0']),\n",
       "             ('17989347', ['1', '1']),\n",
       "             ('17992131', ['0', '0']),\n",
       "             ('17994639', ['1', '1']),\n",
       "             ('17998327', ['1', '1']),\n",
       "             ('18000742', ['1', '1']),\n",
       "             ('18002105', ['1', '1']),\n",
       "             ('18055357', ['1', '1']),\n",
       "             ('18081136', ['1', '1']),\n",
       "             ('18084033', ['1', '1']),\n",
       "             ('18084663', ['1', '1', '1', '1']),\n",
       "             ('18087458', ['0', '0', '0', '0']),\n",
       "             ('18088250', ['1', '1']),\n",
       "             ('18088903', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('18090006', ['0', '0']),\n",
       "             ('18090946', ['1', '1']),\n",
       "             ('18092649', ['0', '0', '0', '0']),\n",
       "             ('18094581', ['1', '1']),\n",
       "             ('18095838', ['0', '0']),\n",
       "             ('18101834', ['1', '1']),\n",
       "             ('18103556', ['1', '1', '1']),\n",
       "             ('18106264', ['0', '0', '0']),\n",
       "             ('18108334', ['0', '0', '0']),\n",
       "             ('18111009', ['1', '1']),\n",
       "             ('18114956', ['0', '0', '0', '0', '0']),\n",
       "             ('18119761', ['1', '1', '1']),\n",
       "             ('18124332', ['0', '0', '0', '0']),\n",
       "             ('18129133', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('18132928', ['1', '1']),\n",
       "             ('18134240', ['0', '0', '0']),\n",
       "             ('18136695', ['0', '0']),\n",
       "             ('18139402', ['1', '1']),\n",
       "             ('18168703', ['1', '1']),\n",
       "             ('18169975', ['1', '1', '1']),\n",
       "             ('18171567', ['0', '0', '0']),\n",
       "             ('18172337', ['0']),\n",
       "             ('18188989', ['1', '1', '1']),\n",
       "             ('18189469', ['0', '0']),\n",
       "             ('18202850', ['0', '0']),\n",
       "             ('18209613', ['1', '1']),\n",
       "             ('18210560', ['1', '1', '1']),\n",
       "             ('18395877', ['0', '0', '0']),\n",
       "             ('18418165', ['1', '1']),\n",
       "             ('18418894', ['1', '1']),\n",
       "             ('18421821', ['1', '1', '1']),\n",
       "             ('18426372', ['1', '1']),\n",
       "             ('18433110', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('18436944', ['1', '1', '1']),\n",
       "             ('18438221', ['1', '1', '1', '1', '1']),\n",
       "             ('18439047', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('18443287',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('18445692',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('18449130',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('18455122', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('18456874', ['1', '1', '1', '1']),\n",
       "             ('18457720', ['1', '1']),\n",
       "             ('18458621', ['0', '0', '0', '1']),\n",
       "             ('18459576', ['1', '1', '1', '1']),\n",
       "             ('18460720', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('18463216', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('18464410', ['1', '1']),\n",
       "             ('18465105', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('18466779',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('18479072', ['1', '1']),\n",
       "             ('18495139', ['1', '1']),\n",
       "             ('18501681', ['1', '1']),\n",
       "             ('18506690', ['1', '1']),\n",
       "             ('18509813', ['0', '0']),\n",
       "             ('18512462', ['1', '1', '1']),\n",
       "             ('18518651', ['1', '1']),\n",
       "             ('18521736', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('18523416', ['1', '1', '1']),\n",
       "             ('18528673', ['1', '1', '1', '1', '1']),\n",
       "             ('18530897', ['1', '1', '1', '1']),\n",
       "             ('18579761', ['1', '1', '1']),\n",
       "             ('18584379', ['0', '0']),\n",
       "             ('18587264', ['0', '0', '0']),\n",
       "             ('18608402', ['0', '0', '0', '0']),\n",
       "             ('18610286', ['1', '1']),\n",
       "             ('18633446', ['0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('18634651', ['1', '1', '1']),\n",
       "             ('18875342',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '0',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('18889794', ['0', '1']),\n",
       "             ('18938190', ['1', '1']),\n",
       "             ('18974973', ['0', '0', '0', '0', '0']),\n",
       "             ('19134408', ['1', '1']),\n",
       "             ('19312361', ['1', '1']),\n",
       "             ('19320844', ['1', '1']),\n",
       "             ('19325308', ['1', '1']),\n",
       "             ('19426032', ['1', '1']),\n",
       "             ('19428077', ['0', '0']),\n",
       "             ('19430259', ['1', '1']),\n",
       "             ('19431396', ['0', '0']),\n",
       "             ('19444915', ['0', '0']),\n",
       "             ('19449473', ['0', '0']),\n",
       "             ('19483898', ['0', '0']),\n",
       "             ('19495749', ['1', '1', '1']),\n",
       "             ('19561066', ['1', '1']),\n",
       "             ('19562492', ['0', '0', '0', '0']),\n",
       "             ('19565850', ['1', '1']),\n",
       "             ('19567776', ['1', '1', '1', '1']),\n",
       "             ('19574488',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('19576595', ['1', '1']),\n",
       "             ('19581830', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('19584478', ['1', '0', '0']),\n",
       "             ('19588076', ['1', '1']),\n",
       "             ('19588837', ['1', '1', '1', '1', '1']),\n",
       "             ('19593156', ['1', '1', '1', '1']),\n",
       "             ('19594436', ['1', '1', '1']),\n",
       "             ('19633151', ['1', '1', '1']),\n",
       "             ('19639693', ['1', '1', '1', '1']),\n",
       "             ('19641083', ['1', '1']),\n",
       "             ('19642419', ['0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('19644303', ['0', '0', '0', '0']),\n",
       "             ('19645606', ['1', '1', '1', '1', '1']),\n",
       "             ('19654452', ['1', '1']),\n",
       "             ('19668057', ['1', '1']),\n",
       "             ('19670913', ['1', '1']),\n",
       "             ('19675460', ['1', '1']),\n",
       "             ('19675944', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('19677013',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('19681522', ['1', '1']),\n",
       "             ('19682225', ['0', '0', '0']),\n",
       "             ('19684006', ['1', '1']),\n",
       "             ('19690656', ['0', '0']),\n",
       "             ('19824999', ['1', '1']),\n",
       "             ('19826640', ['0', '0']),\n",
       "             ('19833781', ['0', '0', '0', '0', '0']),\n",
       "             ('19836746', ['1', '1', '1', '1', '1']),\n",
       "             ('19899934', ['0', '0']),\n",
       "             ('19905802', ['1', '1', '1', '1', '1']),\n",
       "             ('19907118', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('19913726', ['1', '1', '1']),\n",
       "             ('19919405', ['1', '1']),\n",
       "             ('19920359', ['1', '1']),\n",
       "             ('19921825', ['0', '0', '0']),\n",
       "             ('19923153', ['0', '0', '0', '0']),\n",
       "             ('19931853', ['0', '0', '0', '0']),\n",
       "             ('19933687', ['0', '0']),\n",
       "             ('19935607', ['0', '0']),\n",
       "             ('20740778', ['1', '1']),\n",
       "             ('20753041', ['1', '1']),\n",
       "             ('20753795', ['1', '1']),\n",
       "             ('20761384', ['1', '1']),\n",
       "             ('20764559', ['1', '1']),\n",
       "             ('20764994', ['0', '0']),\n",
       "             ('20765538', ['0', '0']),\n",
       "             ('20768635', ['1', '1']),\n",
       "             ('20774066', ['1', '1', '1']),\n",
       "             ('20775200', ['0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('20953850', ['1', '1']),\n",
       "             ('20967164', ['1', '1']),\n",
       "             ('20999190', ['1', '1']),\n",
       "             ('21025935', ['0', '0', '0']),\n",
       "             ('21028946', ['0', '0']),\n",
       "             ('21033221', ['0', '0']),\n",
       "             ('21039805', ['0', '0', '0', '0']),\n",
       "             ('21043415', ['1', '1']),\n",
       "             ('21053389', ['1', '1']),\n",
       "             ('21091887', ['0', '0']),\n",
       "             ('21095344', ['1', '1', '1']),\n",
       "             ('21107153', ['1', '1']),\n",
       "             ('21111072', ['0', '0']),\n",
       "             ('21117927', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('21123247', ['0', '0']),\n",
       "             ('21128206', ['1', '1']),\n",
       "             ('21129295', ['1', '1', '1']),\n",
       "             ('21130129', ['1', '1', '1']),\n",
       "             ('21164643', ['1', '1']),\n",
       "             ('21165675', ['0', '0', '0']),\n",
       "             ('21168455', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('21171069', ['1', '1']),\n",
       "             ('21174312', ['1', '1', '1', '1']),\n",
       "             ('21177245', ['1', '1']),\n",
       "             ('21178464', ['1', '1']),\n",
       "             ('21181465', ['1', '1']),\n",
       "             ('21189817', ['1', '1']),\n",
       "             ('21192208', ['1', '1']),\n",
       "             ('21196634', ['0', '0']),\n",
       "             ('21197679', ['1', '1']),\n",
       "             ('21198776', ['1', '1']),\n",
       "             ('21200315', ['1', '1']),\n",
       "             ('21202747', ['1', '1']),\n",
       "             ('21203879', ['0', '0', '0']),\n",
       "             ('21205257', ['0', '0']),\n",
       "             ('21207566', ['1', '1', '1']),\n",
       "             ('21208236', ['1', '1']),\n",
       "             ('21212090', ['1', '1', '1']),\n",
       "             ('21213380', ['1', '1']),\n",
       "             ('21216695', ['0', '0']),\n",
       "             ('21220651', ['0', '0']),\n",
       "             ('21223303', ['0', '0']),\n",
       "             ('21448034',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('21451498', ['0', '0']),\n",
       "             ('21451951', ['0', '0']),\n",
       "             ('21809133', ['0', '0', '0']),\n",
       "             ('21811991', ['1', '1']),\n",
       "             ('22020694', ['1', '1', '1']),\n",
       "             ('22039863', ['1', '1']),\n",
       "             ('22045469', ['1', '1']),\n",
       "             ('22055394', ['1', '1']),\n",
       "             ('22056915', ['0', '1']),\n",
       "             ('22065344', ['1', '1']),\n",
       "             ('22071144', ['1', '1']),\n",
       "             ('22075419', ['1', '1']),\n",
       "             ('22081867', ['1', '1']),\n",
       "             ('22082820', ['1', '1']),\n",
       "             ('22093592', ['1', '1']),\n",
       "             ('22145101', ['1', '1']),\n",
       "             ('22156789', ['0', '0']),\n",
       "             ('22163425', ['0', '0']),\n",
       "             ('22185347', ['0', '0', '0', '0', '0']),\n",
       "             ('22191575', ['1', '1']),\n",
       "             ('22212469', ['0', '0']),\n",
       "             ('22214572', ['0', '0']),\n",
       "             ('22220816', ['0', '1']),\n",
       "             ('22227448', ['1', '1']),\n",
       "             ('22340371', ['1', '1']),\n",
       "             ('22341979', ['1', '1']),\n",
       "             ('22342992', ['1', '1']),\n",
       "             ('22354548', ['1', '1']),\n",
       "             ('22355368', ['0', '0', '0']),\n",
       "             ('22357271', ['0', '0']),\n",
       "             ('22405737', ['1', '1']),\n",
       "             ('22406560', ['1', '1', '1']),\n",
       "             ('22409255', ['1', '1', '1', '1']),\n",
       "             ('22412556', ['1', '1', '1']),\n",
       "             ('22413793', ['0', '0']),\n",
       "             ('22414120', ['0', '0', '0']),\n",
       "             ('22414988', ['1', '1']),\n",
       "             ('22415546', ['1', '1', '1', '1', '1']),\n",
       "             ('22418252', ['1', '1', '1', '1', '1']),\n",
       "             ('22420787', ['1', '1']),\n",
       "             ('22424184', ['1', '1', '1']),\n",
       "             ('22433851', ['1', '1']),\n",
       "             ('22436273', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22438959', ['1', '1', '1']),\n",
       "             ('22440066', ['1', '1']),\n",
       "             ('22440788', ['1', '1']),\n",
       "             ('22485124', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22487256', ['0', '0', '0']),\n",
       "             ('22488043', ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('22491660', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22494542', ['0', '0']),\n",
       "             ('22496884', ['1', '1', '1']),\n",
       "             ('22504281', ['1', '1']),\n",
       "             ('22505157', ['1', '1', '1']),\n",
       "             ('22509108', ['0', '0']),\n",
       "             ('22517136', ['1', '1', '1']),\n",
       "             ('22518338', ['1', '1', '1']),\n",
       "             ('22520067', ['0', '0']),\n",
       "             ('22522265',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22527921', ['1', '1', '1']),\n",
       "             ('22529723', ['1', '1', '1']),\n",
       "             ('22532653', ['1', '1']),\n",
       "             ('22543227', ['1', '1']),\n",
       "             ('22549634', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22551782', ['0', '0', '0']),\n",
       "             ('22552515', ['1', '1', '1', '1']),\n",
       "             ('22554716', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22556368', ['0', '0']),\n",
       "             ('22566968', ['1', '1']),\n",
       "             ('22570271', ['1', '1', '1']),\n",
       "             ('22572822', ['1', '1', '1']),\n",
       "             ('22582778', ['0', '0']),\n",
       "             ('22586065', ['1', '1']),\n",
       "             ('22588632', ['1', '1', '1', '1']),\n",
       "             ('22590113', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22593041', ['0', '0', '0', '0']),\n",
       "             ('22595831', ['1', '1', '1', '1']),\n",
       "             ('22597074', ['1', '1', '1', '1']),\n",
       "             ('22599537', ['0', '0']),\n",
       "             ('22600912', ['1', '1', '1']),\n",
       "             ('22602768', ['1', '1', '1']),\n",
       "             ('22605025', ['1', '1']),\n",
       "             ('22609303', ['1', '1']),\n",
       "             ('22610156', ['1', '1', '1', '1', '1']),\n",
       "             ('22612897', ['0', '0']),\n",
       "             ('22617339', ['1', '1']),\n",
       "             ('22635923', ['0', '0', '0', '0']),\n",
       "             ('22654384', ['0', '0', '0']),\n",
       "             ('22660748', ['0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('22666360', ['1', '1', '1']),\n",
       "             ('22668995', ['0', '0']),\n",
       "             ('22669748', ['1', '1', '1']),\n",
       "             ('22670585', ['1', '1', '1']),\n",
       "             ('22671741', ['1', '1', '1']),\n",
       "             ('22672483', ['1', '1']),\n",
       "             ('22673639', ['1', '1']),\n",
       "             ('22674245', ['1', '1']),\n",
       "             ('22675289', ['1', '1', '1']),\n",
       "             ('22676110', ['1', '1', '1', '1']),\n",
       "             ('22677968', ['1', '1', '1', '1']),\n",
       "             ('22679661', ['1', '1', '1']),\n",
       "             ('22681025',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('22690080', ['1', '1']),\n",
       "             ('22693893', ['1', '1', '1']),\n",
       "             ('22697014',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('22701327', ['1', '1']),\n",
       "             ('22715979', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22717444',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('22719864', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22721652', ['1', '1', '1']),\n",
       "             ('22722482', ['0', '0', '0', '0', '0']),\n",
       "             ('22723864', ['1', '1', '1']),\n",
       "             ('22724951',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22729134', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22731758', ['1', '1', '1']),\n",
       "             ('22735789', ['1', '1']),\n",
       "             ('22737066', ['0', '0']),\n",
       "             ('22743256', ['0', '0', '0', '0']),\n",
       "             ('22745971', ['1', '1']),\n",
       "             ('22747677', ['0', '0']),\n",
       "             ('22749084', ['1', '1']),\n",
       "             ('22777548', ['1', '1']),\n",
       "             ('22780274', ['0', '0']),\n",
       "             ('22793107', ['1', '1', '1']),\n",
       "             ('22830618', ['1', '1']),\n",
       "             ('22879227', ['0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('22892303', ['1', '1']),\n",
       "             ('22913667', ['1', '1']),\n",
       "             ('22915822', ['1', '1']),\n",
       "             ('22916569', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('22921146', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('22922640', ['1', '1']),\n",
       "             ('22926137', ['1', '1', '1']),\n",
       "             ('22928563',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('22931376', ['1', '1', '1']),\n",
       "             ('22932152', ['1', '1']),\n",
       "             ('22939608', ['1', '1', '1']),\n",
       "             ('22942720', ['1', '1']),\n",
       "             ('22943500', ['1', '1', '1']),\n",
       "             ('22946974', ['1', '1']),\n",
       "             ('22951024', ['1', '1']),\n",
       "             ('22951607', ['1', '1', '1', '1']),\n",
       "             ('22953626', ['1', '1', '1']),\n",
       "             ('22955296', ['1', '1']),\n",
       "             ('22957510', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('22960217', ['1', '1']),\n",
       "             ('23022136', ['1', '1', '1', '1']),\n",
       "             ('23023222', ['1', '1', '1', '1']),\n",
       "             ('23025022', ['1', '1', '1']),\n",
       "             ('23027209', ['1', '1']),\n",
       "             ('23029526', ['1', '1']),\n",
       "             ('23031533', ['1', '1', '1']),\n",
       "             ('23035012',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('23040362', ['1', '1', '1', '1', '1']),\n",
       "             ('23041862', ['0', '0']),\n",
       "             ('23045012', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23050471', ['1', '1', '1']),\n",
       "             ('23063840',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23065506', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('23066679', ['1', '1', '1']),\n",
       "             ('23067476', ['1', '1']),\n",
       "             ('23069118', ['1', '1']),\n",
       "             ('23071667', ['0', '0']),\n",
       "             ('23077555', ['1', '1', '1', '1']),\n",
       "             ('23079989', ['1', '1']),\n",
       "             ('23080812',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('23087000', ['1', '1']),\n",
       "             ('23090423', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23091680', ['1', '1']),\n",
       "             ('23092160', ['1', '1', '1', '1']),\n",
       "             ('23094201', ['1', '1']),\n",
       "             ('23097881', ['0', '0', '0']),\n",
       "             ('23099363', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('23101055', ['1', '1', '1', '1', '1']),\n",
       "             ('23102312', ['1', '1', '1', '1']),\n",
       "             ('23104069', ['1', '1', '1', '1', '1']),\n",
       "             ('23105516',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23107269',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('23110613', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('23111635', ['1', '1', '1']),\n",
       "             ('23132910', ['1', '1']),\n",
       "             ('23137135', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('23140940', ['1', '1']),\n",
       "             ('23141628', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('23178553', ['1', '1']),\n",
       "             ('23180233', ['0', '0']),\n",
       "             ('23183714', ['1', '1']),\n",
       "             ('23185571', ['1', '1']),\n",
       "             ('23192987', ['1', '1', '1', '1']),\n",
       "             ('23195777', ['1', '1']),\n",
       "             ('23202615',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('23205980', ['1', '1']),\n",
       "             ('23206526', ['1', '1', '1']),\n",
       "             ('23210513', ['0', '0']),\n",
       "             ('23216543', ['1', '1', '1', '1']),\n",
       "             ('23223858',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('23229918', ['1', '1']),\n",
       "             ('23238289', ['1', '1']),\n",
       "             ('23246565', ['1', '1']),\n",
       "             ('23248758', ['1', '1']),\n",
       "             ('23249040', ['0', '0', '0']),\n",
       "             ('23255890', ['1', '1', '1']),\n",
       "             ('23278515', ['1', '1', '1', '1', '1']),\n",
       "             ('23283041', ['1', '1', '1']),\n",
       "             ('23284781', ['1', '1', '1', '1']),\n",
       "             ('23625330', ['1', '1']),\n",
       "             ('23630617', ['0', '0', '0']),\n",
       "             ('23713809', ['1', '1']),\n",
       "             ('23772220', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('23773236', ['1', '1']),\n",
       "             ('23781627', ['1', '1']),\n",
       "             ('23782725', ['0', '0']),\n",
       "             ('23787339', ['1', '1']),\n",
       "             ('23788591', ['1', '1']),\n",
       "             ('23792618', ['1', '1']),\n",
       "             ('23794573', ['0', '0']),\n",
       "             ('23801033', ['1', '1', '1']),\n",
       "             ('23803608', ['1', '1', '1']),\n",
       "             ('23807603', ['1', '1']),\n",
       "             ('23809677', ['0', '0', '0']),\n",
       "             ('23812823', ['1', '1', '1']),\n",
       "             ('23815080', ['0', '0']),\n",
       "             ('23815514', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23817027',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23819325', ['1', '1', '1']),\n",
       "             ('23829263', ['1', '1']),\n",
       "             ('23830493', ['1', '1']),\n",
       "             ('23868956', ['1', '1', '1']),\n",
       "             ('23870079', ['1', '1']),\n",
       "             ('23870528', ['1', '1', '1']),\n",
       "             ('23880813', ['1', '1']),\n",
       "             ('23881294', ['1', '1']),\n",
       "             ('23883810', ['1', '1']),\n",
       "             ('23915889', ['1', '1', '1']),\n",
       "             ('23924822', ['1', '1', '1', '1', '1']),\n",
       "             ('23928891', ['1', '1']),\n",
       "             ('23929674', ['1', '1', '1', '1']),\n",
       "             ('23931845', ['0', '0']),\n",
       "             ('23937135', ['1', '1']),\n",
       "             ('23937811', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('23940256', ['1', '1', '1']),\n",
       "             ('23944762', ['1', '1', '1', '1', '1']),\n",
       "             ('23946990', ['0', '0', '0']),\n",
       "             ('23949744', ['1', '1']),\n",
       "             ('23977055', ['0', '0', '0']),\n",
       "             ('23979342', ['1', '1']),\n",
       "             ('23980755',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('23982775', ['1', '1', '1']),\n",
       "             ('23984653', ['1', '1']),\n",
       "             ('23994657',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('23997651', ['1', '1', '1']),\n",
       "             ('23998780', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('24000258', ['1', '1', '1']),\n",
       "             ('24004787', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('24011356', ['1', '1', '1']),\n",
       "             ('24013178', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('24015915', ['1', '1', '1']),\n",
       "             ('24018850', ['0', '0']),\n",
       "             ('24020575',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('24022631', ['1', '1', '1', '1', '1']),\n",
       "             ('24024933', ['1', '1', '1', '1']),\n",
       "             ('24028351', ['1', '1']),\n",
       "             ('24050841', ['0', '0']),\n",
       "             ('24066399', ['0', '0']),\n",
       "             ('24072906', ['1', '1']),\n",
       "             ('24079243', ['0', '0']),\n",
       "             ('24088324', ['1', '1']),\n",
       "             ('24191106', ['1', '1']),\n",
       "             ('24192879', ['1', '1']),\n",
       "             ('24194941', ['1', '1']),\n",
       "             ('24195882', ['1', '1']),\n",
       "             ('24206571', ['1', '1', '1']),\n",
       "             ('24206880', ['1', '1']),\n",
       "             ('24230409', ['1', '1']),\n",
       "             ('24232507', ['1', '1', '1', '1']),\n",
       "             ('24233592', ['1', '1', '1', '1', '1']),\n",
       "             ('24401503', ['1', '1']),\n",
       "             ('24421786', ['0', '0', '0']),\n",
       "             ('24423142', ['0', '0']),\n",
       "             ('24428785', ['1', '1']),\n",
       "             ('24449928', ['1', '1']),\n",
       "             ('24490529', ['1', '1']),\n",
       "             ('24527749', ['0', '0']),\n",
       "             ('24601134', ['1', '1']),\n",
       "             ('24602289', ['1', '1', '1']),\n",
       "             ('24604666', ['1', '1', '1', '1']),\n",
       "             ('24607001', ['0', '0', '0', '0', '1']),\n",
       "             ('24613907', ['0', '0']),\n",
       "             ('24616876',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('24621703', ['1', '1']),\n",
       "             ('24625011', ['0', '0']),\n",
       "             ('24631342', ['0', '0']),\n",
       "             ('24633407', ['1', '1']),\n",
       "             ('24635927', ['1', '1']),\n",
       "             ('24681555', ['1', '1']),\n",
       "             ('24741061', ['1', '1']),\n",
       "             ('24745305', ['1', '1']),\n",
       "             ('24750512', ['1', '1']),\n",
       "             ('24751158', ['1', '1']),\n",
       "             ('25073388', ['0', '0', '0']),\n",
       "             ('25075866', ['1', '1']),\n",
       "             ('25076598', ['0', '0']),\n",
       "             ('25080956', ['1', '1']),\n",
       "             ('25082586', ['1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('25084713', ['1', '1', '1']),\n",
       "             ('25085373', ['1', '1', '1']),\n",
       "             ('25087164', ['1', '1', '1', '1', '1']),\n",
       "             ('25088494', ['1', '1']),\n",
       "             ('25089322',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('25094393', ['1', '1', '1']),\n",
       "             ('25108767', ['1', '1']),\n",
       "             ('25110395', ['1', '1', '1']),\n",
       "             ('25130908', ['1', '1', '1', '1']),\n",
       "             ('25132498', ['1', '1']),\n",
       "             ('25144421', ['1', '1', '1']),\n",
       "             ('25211793', ['1', '1']),\n",
       "             ('25214682', ['1', '1']),\n",
       "             ('25217701', ['1', '1', '1', '1']),\n",
       "             ('25219697', ['1', '1']),\n",
       "             ('25237029', ['1', '1']),\n",
       "             ('25265790', ['1', '1']),\n",
       "             ('25292719', ['1', '1']),\n",
       "             ('25294818', ['1', '1']),\n",
       "             ('25297338', ['0', '0', '0', '0']),\n",
       "             ('25298891', ['1', '1']),\n",
       "             ('25300371', ['1', '1']),\n",
       "             ('25301743', ['1', '1']),\n",
       "             ('25303688', ['0', '0', '0', '0']),\n",
       "             ('25304948', ['0', '0', '0', '0', '0']),\n",
       "             ('25308272', ['1', '1']),\n",
       "             ('25309389', ['0', '0', '0', '0']),\n",
       "             ('25311007', ['1', '1', '1']),\n",
       "             ('25403114', ['1', '1']),\n",
       "             ('25408189', ['0', '0']),\n",
       "             ('25410266', ['1', '1']),\n",
       "             ('25412911', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('25450980', ['1', '1']),\n",
       "             ('25460772', ['1', '1']),\n",
       "             ('25464310', ['0', '0']),\n",
       "             ('25466553', ['0', '0']),\n",
       "             ('25470546', ['1', '1']),\n",
       "             ('25476871', ['1', '1']),\n",
       "             ('25480872', ['1', '1', '1']),\n",
       "             ('25492783', ['1', '1']),\n",
       "             ('25495138', ['1', '1']),\n",
       "             ('25506518', ['1', '1', '1']),\n",
       "             ('25511130', ['1', '1']),\n",
       "             ('25519541', ['1', '1']),\n",
       "             ('25525489', ['1', '1']),\n",
       "             ('25531673', ['1', '1']),\n",
       "             ('25532237', ['1', '1']),\n",
       "             ('25539631', ['1', '1', '1']),\n",
       "             ('25560837', ['1', '1']),\n",
       "             ('25565026', ['1', '1']),\n",
       "             ('25569680', ['1', '1']),\n",
       "             ('25572582', ['1', '1']),\n",
       "             ('25576087',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('25582536', ['1', '1', '1', '1']),\n",
       "             ('25585035', ['1', '1']),\n",
       "             ('25586338', ['1', '1']),\n",
       "             ('25587436', ['0', '0', '0', '0']),\n",
       "             ('25589256', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('25591200', ['1', '1']),\n",
       "             ('25593452', ['0', '0']),\n",
       "             ('25640835', ['1', '1', '1', '1']),\n",
       "             ('25641754', ['1', '1', '1', '1', '1']),\n",
       "             ('25643483', ['1', '1', '1', '1']),\n",
       "             ('25678422', ['1', '1']),\n",
       "             ('25687141', ['1', '1', '1', '1', '1']),\n",
       "             ('25697952', ['1', '1', '1', '1', '1']),\n",
       "             ('25699291', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('25702777', ['1', '1', '1']),\n",
       "             ('25704446', ['1', '1', '1']),\n",
       "             ('25707129', ['1', '1', '1', '1', '1']),\n",
       "             ('25708642', ['0', '0', '0']),\n",
       "             ('25709737', ['1', '1']),\n",
       "             ('25714110', ['1', '1']),\n",
       "             ('25810924', ['1', '1', '1', '1', '1']),\n",
       "             ('25813676', ['1', '1']),\n",
       "             ('25895422', ['1', '1', '1']),\n",
       "             ('25901750', ['1', '1']),\n",
       "             ('25902761', ['1', '1', '1']),\n",
       "             ('25905421', ['1', '1', '1', '1']),\n",
       "             ('25906860', ['1', '1', '1', '1']),\n",
       "             ('25908396', ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('25910191', ['1', '1', '1']),\n",
       "             ('25911071', ['1', '1', '1', '1']),\n",
       "             ('25917194',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('25919677', ['0', '0', '0', '0']),\n",
       "             ('25920729', ['1', '1', '1', '1']),\n",
       "             ('25923011', ['0', '0', '0']),\n",
       "             ('25925187', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('25980588', ['1', '1', '1']),\n",
       "             ('26131046', ['1', '1', '1']),\n",
       "             ('26131985', ['1', '1']),\n",
       "             ('26132336', ['0', '0', '0', '0']),\n",
       "             ('26139378', ['1', '1']),\n",
       "             ('26140544', ['1', '1']),\n",
       "             ('26150584', ['1', '1']),\n",
       "             ('26156512', ['1', '1']),\n",
       "             ('26160105', ['1', '1']),\n",
       "             ('26164079', ['1', '1']),\n",
       "             ('26278175', ['0', '0', '0']),\n",
       "             ('26283020', ['0', '0']),\n",
       "             ('26283694', ['0', '0']),\n",
       "             ('26290726', ['1', '1']),\n",
       "             ('26293877', ['1', '1']),\n",
       "             ('26300480', ['0', '0', '0']),\n",
       "             ('26311014', ['1', '1', '1', '1']),\n",
       "             ('26324753', ['1', '1', '1', '1']),\n",
       "             ('26326222', ['1', '1', '1', '1', '1']),\n",
       "             ('26331185', ['1', '1']),\n",
       "             ('26335739', ['1', '1']),\n",
       "             ('26337618', ['1', '1', '1', '1']),\n",
       "             ('26338972', ['1', '1']),\n",
       "             ('26342529', ['1', '1']),\n",
       "             ('26343805', ['0', '0']),\n",
       "             ('26344493', ['1', '1']),\n",
       "             ('26379837', ['1', '1', '1', '1', '1']),\n",
       "             ('26388548', ['1', '1']),\n",
       "             ('26390755', ['0', '0']),\n",
       "             ('26392946', ['1', '1', '1', '1']),\n",
       "             ('26394982', ['1', '1', '1']),\n",
       "             ('26400967', ['0', '0', '0']),\n",
       "             ('26402623', ['1', '1', '1', '1']),\n",
       "             ('26406382', ['0', '0']),\n",
       "             ('26408660', ['1', '1', '1', '1']),\n",
       "             ('26410624', ['1', '1', '1']),\n",
       "             ('26474305', ['0', '0']),\n",
       "             ('26484668', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('26497014', ['1', '1']),\n",
       "             ('26498890', ['0', '0']),\n",
       "             ('26500958', ['1', '1']),\n",
       "             ('26504002', ['0', '0']),\n",
       "             ('26504839', ['1', '1', '1']),\n",
       "             ('26506851', ['1', '1']),\n",
       "             ('26507446', ['1', '1']),\n",
       "             ('26512217', ['1', '1']),\n",
       "             ('26512760', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('26516530', ['1', '1', '1']),\n",
       "             ('26518455', ['1', '1', '1']),\n",
       "             ('26520564', ['0', '0', '0']),\n",
       "             ('26541088', ['1', '1']),\n",
       "             ('26553234', ['1', '1']),\n",
       "             ('26613158', ['1', '1']),\n",
       "             ('26620244', ['1', '1', '1']),\n",
       "             ('26625334', ['1', '1']),\n",
       "             ('26627906', ['1', '1', '1']),\n",
       "             ('26629842', ['0', '0', '0', '0', '0']),\n",
       "             ('26633178', ['1', '1']),\n",
       "             ('26635779', ['1', '1', '1']),\n",
       "             ('26638677', ['0', '0', '0', '0']),\n",
       "             ('26640900', ['1', '1', '1']),\n",
       "             ('26642679', ['1', '1']),\n",
       "             ('26643526', ['1', '1']),\n",
       "             ('26646503', ['1', '1', '1', '1']),\n",
       "             ('26649421', ['1', '1', '1']),\n",
       "             ('26650365', ['1', '1', '1']),\n",
       "             ('26652930', ['1', '1', '1']),\n",
       "             ('26654697', ['1', '1']),\n",
       "             ('26657017', ['0', '0', '0']),\n",
       "             ('26658299', ['1', '1', '1']),\n",
       "             ('26661500', ['1', '1']),\n",
       "             ('26664053', ['0', '0']),\n",
       "             ('26665099', ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('26669791', ['1', '1', '1', '1', '1']),\n",
       "             ('26670960', ['1', '1']),\n",
       "             ('26671590', ['1', '1', '1']),\n",
       "             ('26672315', ['1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('26676419', ['0', '0']),\n",
       "             ('26678128', ['0', '0']),\n",
       "             ('26683328', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('26687443',\n",
       "              ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('26692128', ['1', '1', '1']),\n",
       "             ('26695761', ['0', '0']),\n",
       "             ('26696783', ['1', '1', '1']),\n",
       "             ('26701805', ['0', '0']),\n",
       "             ('26703553', ['1', '1']),\n",
       "             ('26705391', ['1', '1', '1', '1']),\n",
       "             ('26706228', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('26714535', ['1', '1']),\n",
       "             ('26715393', ['1', '1']),\n",
       "             ('26716297', ['1', '1']),\n",
       "             ('26718872', ['1', '1', '1']),\n",
       "             ('26770986', ['1', '1']),\n",
       "             ('26772080', ['0', '0', '0']),\n",
       "             ('26785020', ['1', '1']),\n",
       "             ('26787544', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('26789641',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('26798092', ['1', '1', '1']),\n",
       "             ('26800009', ['0', '0', '0']),\n",
       "             ('26800799', ['1', '1']),\n",
       "             ('26801527', ['0', '0', '0']),\n",
       "             ('26802915',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('26805436', ['0', '0']),\n",
       "             ('26806000', ['0', '0', '0', '0', '0']),\n",
       "             ('26833161', ['1', '1', '1']),\n",
       "             ('26841390', ['1', '1']),\n",
       "             ('26857599', ['0', '0']),\n",
       "             ('26859465', ['0', '0', '0']),\n",
       "             ('26861205', ['1', '1']),\n",
       "             ('26862153', ['0', '0']),\n",
       "             ('26862950', ['0', '0', '0', '0', '0']),\n",
       "             ('26865040', ['0', '0', '0']),\n",
       "             ('26870502', ['1', '1', '1', '1']),\n",
       "             ('26873397', ['1', '1', '1']),\n",
       "             ('26877981', ['1', '1']),\n",
       "             ('26878967', ['1', '1']),\n",
       "             ('26881131', ['1', '1']),\n",
       "             ('26887829', ['0', '0']),\n",
       "             ('26889482', ['1', '1', '1']),\n",
       "             ('26891382', ['0', '0', '0']),\n",
       "             ('26893912', ['0', '0', '0']),\n",
       "             ('26894974', ['1', '1', '1']),\n",
       "             ('26895957', ['1', '1', '1']),\n",
       "             ('26898891', ['0', '0']),\n",
       "             ('26910207', ['0', '0']),\n",
       "             ('26910993', ['1', '1']),\n",
       "             ('26916043', ['1', '1', '0', '0', '0']),\n",
       "             ('26922227', ['1', '1']),\n",
       "             ('26929814', ['1', '1']),\n",
       "             ('26939992', ['1', '1']),\n",
       "             ('26941223', ['1', '1', '1', '1', '1']),\n",
       "             ('26947089', ['1', '1']),\n",
       "             ('26948118',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('26951317', ['1', '1']),\n",
       "             ('26952363', ['1', '1', '1']),\n",
       "             ('26955377', ['1', '1', '1']),\n",
       "             ('26958393', ['1', '1']),\n",
       "             ('26967163', ['1', '1']),\n",
       "             ('26968506', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27021425', ['1', '1']),\n",
       "             ('27026804', ['0', '0', '0']),\n",
       "             ('27029393', ['0', '0']),\n",
       "             ('27030951', ['1', '1']),\n",
       "             ('27033321', ['1', '1']),\n",
       "             ('27043883', ['1', '1']),\n",
       "             ('27045504', ['0', '0']),\n",
       "             ('27050834', ['1', '1', '1', '1']),\n",
       "             ('27054179', ['1', '1']),\n",
       "             ('27054832', ['1', '1']),\n",
       "             ('27056097', ['1', '1', '1']),\n",
       "             ('27057939', ['0', '0']),\n",
       "             ('27058525', ['1', '1']),\n",
       "             ('27060536', ['1', '1']),\n",
       "             ('27065117', ['0', '0', '0']),\n",
       "             ('27068592', ['0', '0']),\n",
       "             ('27074049', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27075975', ['0', '0', '0', '0', '0', '0']),\n",
       "             ('27078560', ['0', '0', '0', '0']),\n",
       "             ('27084303', ['0', '0', '0']),\n",
       "             ('27085338', ['0', '0', '0']),\n",
       "             ('27086367',\n",
       "              ['0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('27089325', ['0', '0', '0']),\n",
       "             ('27091167', ['0', '0', '0']),\n",
       "             ('27104302', ['1', '1']),\n",
       "             ('27110827', ['1', '1']),\n",
       "             ('27112693', ['1', '1', '1', '1', '1']),\n",
       "             ('27117672', ['0', '0', '0', '0', '0']),\n",
       "             ('27119619', ['1', '1', '1']),\n",
       "             ('27120359', ['0', '0']),\n",
       "             ('27123234', ['0', '0', '0', '0', '0']),\n",
       "             ('27125325', ['0', '0']),\n",
       "             ('27127596', ['1', '1']),\n",
       "             ('27145885', ['1', '1', '1']),\n",
       "             ('27174996', ['1', '1']),\n",
       "             ('27188027', ['1', '1']),\n",
       "             ('27188885', ['0', '0']),\n",
       "             ('27214141', ['0', '0']),\n",
       "             ('27222059', ['0', '0']),\n",
       "             ('27226407', ['1', '1']),\n",
       "             ('27227143', ['1', '1']),\n",
       "             ('27228910', ['1', '1', '1']),\n",
       "             ('27229977', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27230380', ['0', '0']),\n",
       "             ('27230983', ['0', '0', '0', '0']),\n",
       "             ('27236603', ['1', '1', '1', '1']),\n",
       "             ('27237893', ['1', '1']),\n",
       "             ('27240025',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('27241644', ['1', '1']),\n",
       "             ('27246556', ['1', '1']),\n",
       "             ('27247770', ['1', '1']),\n",
       "             ('27251367', ['1', '1']),\n",
       "             ('27252843', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27255675',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('27258265', ['1', '1', '1']),\n",
       "             ('27261517', ['0', '0', '0']),\n",
       "             ('27262569', ['0', '0']),\n",
       "             ('27264501', ['1', '1', '1', '1']),\n",
       "             ('27266472', ['1', '1', '1', '1']),\n",
       "             ('27269938', ['0', '0']),\n",
       "             ('27271926', ['1', '1']),\n",
       "             ('27274720', ['0', '0']),\n",
       "             ('27276761', ['0', '0']),\n",
       "             ('27283289', ['1', '1', '1', '1', '1']),\n",
       "             ('27286733', ['1', '1']),\n",
       "             ('27309798', ['1', '1', '1']),\n",
       "             ('27368219', ['0', '0', '0', '0']),\n",
       "             ('27369467', ['0', '0', '0', '0', '0']),\n",
       "             ('27370273', ['0', '0']),\n",
       "             ('27372397', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27373562', ['1', '1']),\n",
       "             ('27377569', ['1', '1', '1']),\n",
       "             ('27380713', ['1', '1', '1', '1']),\n",
       "             ('27381837', ['0', '0']),\n",
       "             ('27383090', ['1', '1', '1', '1']),\n",
       "             ('27384280',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('27386446',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('27395937', ['1', '1', '1']),\n",
       "             ('27397352', ['1', '1']),\n",
       "             ('27398512', ['0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('27400316',\n",
       "              ['1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1',\n",
       "               '1']),\n",
       "             ('27403571', ['1', '1']),\n",
       "             ('27406436', ['1', '1']),\n",
       "             ('27409143', ['0', '0', '1']),\n",
       "             ('27410371', ['1', '1']),\n",
       "             ('27416593', ['1', '1']),\n",
       "             ('27419790', ['0', '0']),\n",
       "             ('27421632', ['1', '1']),\n",
       "             ('27424748', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('27430078', ['1', '1', '1', '1']),\n",
       "             ('27431791', ['1', '1']),\n",
       "             ('27432985', ['1', '1']),\n",
       "             ('27433921', ['1', '1', '1', '1']),\n",
       "             ('27434730', ['1', '1']),\n",
       "             ('27435820', ['1', '1', '1', '1']),\n",
       "             ('27436778', ['1', '1', '1', '1', '1']),\n",
       "             ('27438353', ['1', '1', '1', '1']),\n",
       "             ('27440291', ['1', '1', '1', '1', '1']),\n",
       "             ('27442333', ['1', '1', '1', '1', '1', '1']),\n",
       "             ('27445437', ['1', '1']),\n",
       "             ('27446868', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('27450461', ['1', '1']),\n",
       "             ('27453771', ['1', '1']),\n",
       "             ('27455202', ['1', '1']),\n",
       "             ('27458958', ['1', '1']),\n",
       "             ('27463890', ['0', '0', '0', '0', '0']),\n",
       "             ('27471971', ['1', '1', '1', '1']),\n",
       "             ('27472988', ['1', '1']),\n",
       "             ('27473772', ['1', '1']),\n",
       "             ('27475631', ['1', '1', '1']),\n",
       "             ('27477220', ['1', '1', '1']),\n",
       "             ('27479173', ['0', '0']),\n",
       "             ('27495276', ['0', '0']),\n",
       "             ('27500696', ['1', '1', '1', '1']),\n",
       "             ('27502691', ['1', '1', '1']),\n",
       "             ('27504063', ['1', '1', '1']),\n",
       "             ('27506926', ['0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('27510846', ['0', '0', '0']),\n",
       "             ('27511667', ['1', '1', '1']),\n",
       "             ('27512789', ['1', '1']),\n",
       "             ('27515122', ['0', '0']),\n",
       "             ('27525210', ['1', '1', '1']),\n",
       "             ('27527005', ['1', '1']),\n",
       "             ('27530913', ['1', '1']),\n",
       "             ('27532474', ['1', '1', '1']),\n",
       "             ('27535231', ['1', '1']),\n",
       "             ('27539742', ['1', '1', '1']),\n",
       "             ('27542843', ['1', '1']),\n",
       "             ('27545209',\n",
       "              ['0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0',\n",
       "               '0']),\n",
       "             ('27551790', ['1', '1']),\n",
       "             ('27561081', ['0', '0', '0', '0', '0', '0', '0', '0']),\n",
       "             ('27563613', ['1', '1']),\n",
       "             ('27566498', ['1', '1', '1']),\n",
       "             ('27567545', ['1', '1']),\n",
       "             ('27573745', ['1', '1']),\n",
       "             ('27574804', ['1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('27579150', ['1', '1']),\n",
       "             ('27583037',\n",
       "              ['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1']),\n",
       "             ('27589981', ['1', '1', '1']),\n",
       "             ('27600516', ['1', '1', '1', '1']),\n",
       "             ('27608295', ['1', '1']),\n",
       "             ('27622616', ['1', '1', '1']),\n",
       "             ('27631711', ['1', '1']),\n",
       "             ...])"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapBlock"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<map at 0x7f7290016128>"
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 26,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 8,\n",
       " 5,\n",
       " 2,\n",
       " 7,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 10,\n",
       " 3,\n",
       " 11,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 5,\n",
       " 6,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 7,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 4,\n",
       " 2,\n",
       " 5,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 8,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 6,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 5,\n",
       " 3,\n",
       " 4,\n",
       " 6,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 1,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 3,\n",
       " 5,\n",
       " 8,\n",
       " 18,\n",
       " 17,\n",
       " 19,\n",
       " 6,\n",
       " 4,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 7,\n",
       " 8,\n",
       " 2,\n",
       " 7,\n",
       " 15,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 9,\n",
       " 3,\n",
       " 5,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 8,\n",
       " 3,\n",
       " 19,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 4,\n",
       " 12,\n",
       " 2,\n",
       " 6,\n",
       " 3,\n",
       " 2,\n",
       " 5,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 7,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 13,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 5,\n",
       " 2,\n",
       " 5,\n",
       " 10,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 7,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 6,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 15,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 5,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 9,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 10,\n",
       " 3,\n",
       " 10,\n",
       " 10,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 11,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 7,\n",
       " 3,\n",
       " 4,\n",
       " 9,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 8,\n",
       " 4,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 8,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 4,\n",
       " 3,\n",
       " 13,\n",
       " 2,\n",
       " 3,\n",
       " 19,\n",
       " 2,\n",
       " 7,\n",
       " 13,\n",
       " 7,\n",
       " 3,\n",
       " 5,\n",
       " 3,\n",
       " 12,\n",
       " 10,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 9,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 8,\n",
       " 6,\n",
       " 2,\n",
       " 3,\n",
       " 14,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 14,\n",
       " 5,\n",
       " 2,\n",
       " 7,\n",
       " 3,\n",
       " 12,\n",
       " 6,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 37,\n",
       " 2,\n",
       " 10,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 3,\n",
       " 6,\n",
       " 5,\n",
       " 4,\n",
       " 5,\n",
       " 11,\n",
       " 21,\n",
       " 6,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 13,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 4,\n",
       " 21,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 5,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 8,\n",
       " 11,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 5,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 9,\n",
       " 3,\n",
       " 5,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 23,\n",
       " 3,\n",
       " 2,\n",
       " 29,\n",
       " 3,\n",
       " 7,\n",
       " 3,\n",
       " 7,\n",
       " 3,\n",
       " 9,\n",
       " 3,\n",
       " 2,\n",
       " 11,\n",
       " 5,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 11,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 9,\n",
       " 3,\n",
       " 3,\n",
       " 5,\n",
       " 2,\n",
       " 15,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 21,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 5,\n",
       " 4,\n",
       " 2,\n",
       " 5,\n",
       " 5,\n",
       " 7,\n",
       " 3,\n",
       " 3,\n",
       " 5,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 4,\n",
       " 10,\n",
       " 3,\n",
       " 4,\n",
       " 14,\n",
       " 4,\n",
       " 4,\n",
       " 3,\n",
       " 7,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 10,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 8,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 11,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 7,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 23,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 12,\n",
       " 2,\n",
       " 5,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 3,\n",
       " 4,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 11,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 6,\n",
       " 6,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 12,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 5,\n",
       " 3,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 6,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 22,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 6,\n",
       " 12,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 4,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 2,\n",
       " 6,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 4,\n",
       " 13,\n",
       " 13,\n",
       " 3,\n",
       " 2,\n",
       " 7,\n",
       " 16,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 7,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 2,\n",
       " 4,\n",
       " 5,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 2,\n",
       " 7,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 2,\n",
       " 5,\n",
       " 4,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 4,\n",
       " 3,\n",
       " 3,\n",
       " 8,\n",
       " 3,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 13,\n",
       " 2,\n",
       " 8,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " 2,\n",
       " 7,\n",
       " 2,\n",
       " 11,\n",
       " 3,\n",
       " 4,\n",
       " 2,\n",
       " 3,\n",
       " 2,\n",
       " ...]"
      ]
     },
     "execution_count": 334,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "QAN 3\n"
     ]
    }
   ],
   "source": [
    "\n",
    "QAN50 = 0\n",
    "for size in sorted([len(val) for val in hapBlock.values()], reverse=True):\n",
    "    QAN50 += size\n",
    "    \n",
    "    if QAN50 > phased/2:\n",
    "        print ('QAN %s' %size)\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9469"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4734.5"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
